

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Change log &mdash; GTDB-Tk 2.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/changelog.html" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="_static/custom.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="References" href="references.html" />
    <link rel="prev" title="Example" href="examples/classify_wf.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1"><a class="reference internal" href="commands/index.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Change log</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#id1">2.5.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id5">2.4.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id7">2.4.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id12">2.3.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id16">2.3.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id23">2.2.6</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id25">2.2.5</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id27">2.2.5</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id29">2.2.4</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id31">2.2.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id32">2.2.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id34">2.2.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id36">2.2.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id46">2.1.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id50">2.1.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id54">2.0.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id61">1.7.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id66">1.6.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id69">1.5.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id72">1.5.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id74">1.4.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id77">1.4.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id80">1.4.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id87">1.3.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id88">1.2.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id96">1.1.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id105">1.0.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id106">1.0.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id107">0.3.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id108">0.3.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id109">0.3.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id110">0.3.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id111">0.2.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id112">0.1.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id113">0.1.0</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Change log</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/changelog.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="change-log">
<h1>Change log<a class="headerlink" href="#change-log" title="Permalink to this headline">¶</a></h1>
<div class="section" id="id1">
<h2>2.5.0<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>Major Changes:
* GTDB-Tk now uses <strong>Skani exclusively</strong> for genome clustering, replacing the previous Mash/Skani hybrid approach. This change simplifies the CLI and removes the dependency on Mash, streamlining installation and execution.</p>
<p>Bug Fixes:
* (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/644">#644</a>), (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/641">#641</a>) Fixed compatibility with recent versions of NumPy (≥1.24), which removed the <code class="docutils literal notranslate"><span class="pre">tostring()</span></code> method from <code class="docutils literal notranslate"><span class="pre">numpy.ndarray</span></code>.
* (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/650">#650</a>) Update CLI with an up-to-date taxon.</p>
</div>
<div class="section" id="id5">
<h2>2.4.1<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:
* (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/630">#630</a>) Fixed SyntaxWarning in Python 3.12 by using raw strings for regex in HMMResultsIO.py</p>
</div>
<div class="section" id="id7">
<h2>2.4.0<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/576">#576</a>) When all genomes fail the prodigal step in the classify_wf, The</p></li>
</ul>
<p>bac120 summary file is still produced with the all failed genomes listed as ‘Unclassified’
* (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/573">#573</a>) When running the 3 classify steps independently, a genome can be filtered out in the align
step but still be classified in the identify step. To avoid duplication of row, the genome is classified with a warning.
* (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/540">#540</a>) Empty files are skipped during the sketch step of Mash,
they are then catch in the prodigal step and are returned as ‘Unclassified’
* (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/549">#549</a>) : <cite>–force</cite> has been modified to deal with #540. Prodigal
wasn’t returning the empty files as failed genomes, it was only skipping them. These genomes are now returned in the summary file and flagged as Unclassified.</p>
<p>Major Changes:</p>
<ul>
<li><p>FastANI has been replaced by skani as the primary tool for computing Average Nucleotide Identity (ANI).Users may notice slight variations in the results compared to those obtained using FastANI.</p></li>
<li><p>In the generated <cite>summary.tsv</cite> files, several columns have been renamed for clarity and consistency. The following columns have been affected:</p>
<blockquote>
<div><ul class="simple">
<li><p>“<cite>fastani_reference</cite>” column has been renamed to “<cite>closest_genome_reference</cite>”.</p></li>
<li><p>“<cite>fastani_reference_radius</cite>” column has been renamed to “<cite>closest_genome_reference_radius</cite>”.</p></li>
<li><p>“<cite>fastani_taxonomy</cite>” column has been renamed to “<cite>closest_genome_taxonomy</cite>”.</p></li>
<li><p>“<cite>fastani_ani</cite>” column has been renamed to “<cite>closest_genome_ani</cite>”.</p></li>
<li><p>“<cite>fastani_af</cite>” column has been renamed to “<cite>closest_genome_af</cite>”.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<blockquote>
<div><p>These changes have been implemented to improve the readability and understanding of the data within the <cite>summary.tsv</cite> files. Users should update their scripts or processes accordingly to reflect these renamed column headers.</p>
</div></blockquote>
</div>
<div class="section" id="id12">
<h2>2.3.2<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/528">#528</a>) (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/529">#529</a>) setup.py has been modified to restrict pydantic version to &gt;=1.9.2 and &lt; 2.0a1</p></li>
</ul>
<p>Minor Changes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/526">#526</a>) change captures the Mash stderr in a separate buffer ( Thanks &#64;wasade for your contribution)</p></li>
</ul>
</div>
<div class="section" id="id16">
<h2>2.3.0<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/508">#508</a>) (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/509">#509</a>) If <strong>ALL</strong> genomes for a specific domain are either filtered out or classified with ANI they are now reported in the summary file.</p></li>
</ul>
<p>Minor changes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/491">#491</a>) (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/498">#498</a>) Allow GTDB-Tk to show <code class="docutils literal notranslate"><span class="pre">--help</span></code> and <code class="docutils literal notranslate"><span class="pre">-v</span></code> without <code class="docutils literal notranslate"><span class="pre">GTDBTK_DATA_PATH</span></code> being set.
* WARNING: This is a breaking change if you are importing GTDB-Tk as a library and importing values from <code class="docutils literal notranslate"><span class="pre">gtdbtk.config.config</span></code>, instead you need to import as <code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">gtdbtk.config.common</span> <span class="pre">import</span> <span class="pre">CONFIG</span></code> then access values via <code class="docutils literal notranslate"><span class="pre">CONFIG.&lt;var&gt;</span></code></p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/508">#508</a>) Mash distance is changed from 0.1 to 0.15 . This is will increase the number of FastANI comparisons but will cover cases wheere genomes have a larger Mash distance but a small ANI.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/497">#497</a>) Add a <code class="docutils literal notranslate"><span class="pre">convert_to_species</span></code> function is GTDB-Tk to replace GCA/GCF ids with their GTDB species name</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">--db_version</span></code> flag to <code class="docutils literal notranslate"><span class="pre">check_install</span></code> to check the version of previous GTDB-Tk packages.</p></li>
</ul>
</div>
<div class="section" id="id23">
<h2>2.2.6<a class="headerlink" href="#id23" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/493">#493</a>) Fix issue with –full-tree flag (related to skipping ANI steps)</p></li>
</ul>
<p>Minor changes:</p>
<ul class="simple">
<li><p>Change URL for documentation to ‘<a class="reference external" href="https://ecogenomics.github.io/GTDBTk/installing/index.html">https://ecogenomics.github.io/GTDBTk/installing/index.html</a>’</p></li>
<li><p>Improve portability of the ANI_screen step by regenerating the paths of reference genomes in the current filesystem for mash_db.msh</p></li>
</ul>
</div>
<div class="section" id="id25">
<h2>2.2.5<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>gtdbtk.json is now reset when the pipeline is re run and the status of ani_Screen is not ‘complete’</p></li>
</ul>
<p>Minor changes:</p>
<ul class="simple">
<li><dl class="simple">
<dt>When using ‘–genes’ , ANI steps are skipped and Warnings are raised to the user to</dt><dd><p>inform them that classification is less accurate.</p>
</dd>
</dl>
</li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/486">#486</a>) Environment variables can be used in GTDBTK_DATA_PATH</p></li>
<li><p>‘is_consistent’ function in ‘mash.py’ compares only the filenames, not the full paths</p></li>
<li><p>Add cutoff arguments to PfamScan ( Thanks &#64;AroneyS for the contribution)</p></li>
</ul>
</div>
<div class="section" id="id27">
<h2>2.2.5<a class="headerlink" href="#id27" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>gtdbtk.json is now reset when the pipeline is re run and the status of ani_Screen is not ‘complete’</p></li>
</ul>
<p>Minor changes:</p>
<ul class="simple">
<li><p>When using ‘–genes’ , ANI steps are skipped and Warnings are raised to the user to inform them that classification is less accurate.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/486">#486</a>) Environment variables can be used in GTDBTK_DATA_PATH</p></li>
<li><p>‘is_consistent’ function in ‘mash.py’ compares only the filenames, not the full paths</p></li>
<li><p>Add cutoff arguments to PfamScan ( Thanks &#64;AroneyS for the contribution)</p></li>
</ul>
</div>
<div class="section" id="id29">
<h2>2.2.4<a class="headerlink" href="#id29" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/475">#475</a>) If all genomes are classified using ANI, Tk will skip the identify step and align steps</p></li>
</ul>
<p>Minor changes:</p>
<ul class="simple">
<li><p>Add hidden ‘–skip_pplacer’ flag to skip pplacer step ( useful for debugging)</p></li>
<li><p>Improve documentation</p></li>
<li><p>Convert stage_logger to a Singleton class</p></li>
<li><p>Use existing ANI results if available</p></li>
</ul>
</div>
<div class="section" id="id31">
<h2>2.2.3<a class="headerlink" href="#id31" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>Fix prodigal_fail_counter issue</p></li>
</ul>
</div>
<div class="section" id="id32">
<h2>2.2.2<a class="headerlink" href="#id32" title="Permalink to this headline">¶</a></h2>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/471">#471</a>) Fix Pplacer issue</p></li>
</ul>
</div>
<div class="section" id="id34">
<h2>2.2.1<a class="headerlink" href="#id34" title="Permalink to this headline">¶</a></h2>
<p>Build:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/470">#470</a>) Add missing Pydantic dependency.</p></li>
</ul>
</div>
<div class="section" id="id36">
<h2>2.2.0<a class="headerlink" href="#id36" title="Permalink to this headline">¶</a></h2>
<p>Minor changes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/433">#433</a>) Added additional checks to ensure that the <cite>–outgroup_taxon</cite> cannot be set to a domain (<cite>root</cite>, <cite>de_novo_wf</cite>).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/459">#459</a> / <a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/462">#462</a> ) Fix deprecated np.bool in prodigal_biolib.py. Special thanks to &#64;neoformit for his contribution.</p></li>
<li><p>(<a class="reference external" href="http://github.com/Ecogenomics/GTDBTk/issues/466">#466</a>) RED value has been rounded to 5 decimals after the comma.</p></li>
<li><p>(<a class="reference external" href="http://github.com/Ecogenomics/GTDBTk/issues/451">#451</a>) Extra checks have been added when Prodigal fails.</p></li>
<li><p>(<a class="reference external" href="http://github.com/Ecogenomics/GTDBTk/issues/448">#448</a>) Warning has been added when all the genomes are filtered out and not classified.</p></li>
</ul>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/420">#420</a>) Fixed an issue where GTDB-Tk might hang when classifying TIGRFAM markers (<cite>identify</cite>, <cite>classify_wf</cite>, <cite>de_novo_wf</cite>). Special thanks to &#64;lfenske-93 and &#64;sjaenick for their contribution.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/428">#428</a>) Fixed an issue where the <cite>–gtdbtk_classification_file</cite> would raise an error trying to read the <cite>classify</cite> summary (<cite>root</cite>, <cite>de_novo_wf</cite>).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/439">#439</a>) Fix the pipeline when using protein files instead of nucleotide files. symlink uses absolute path instead.</p></li>
</ul>
</div>
<div class="section" id="id46">
<h2>2.1.1<a class="headerlink" href="#id46" title="Permalink to this headline">¶</a></h2>
<p>Documentation:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/410">#410</a>) Add documentation for <cite>convert_to_itol</cite></p></li>
</ul>
<p>Bug Fixes:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/399">#399</a>) Fix <cite>–genes</cite> option attempting to create a directory.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/400">#400</a>) Updated contig.py to fix inconsistent pplacer paths causing the program to crash.</p></li>
</ul>
</div>
<div class="section" id="id50">
<h2>2.1.0<a class="headerlink" href="#id50" title="Permalink to this headline">¶</a></h2>
<p>Major changes:</p>
<ul class="simple">
<li><p>GTDB-TK now uses a <strong>divide-and-conquer</strong> approach where the bacterial reference tree is split into multiple <strong>class</strong>-level subtrees. This reduces the memory requirements of GTDB-Tk from <strong>320 GB</strong> of RAM when using the full GTDB R07-RS207 reference tree to approximately <strong>55 GB</strong>. A manuscript describing this approach is in preparation. If you wish to continue using the full GTDB reference tree use the <cite>–full-tree</cite> flag. This is the main change from v2.0.0. The split tree approach has been modified from order-level trees to class-level trees to resolve specific classification issues (see <a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/383">#383</a>).</p></li>
<li><p>Genomes that cannot be assigned to a domain (e.g. genomes with no bacterial or archaeal markers or genomes with no genes called by Prodigal) are now reported in the <cite>gtdbtk.bac120.summary.tsv</cite> as ‘Unclassified’</p></li>
<li><p>Genomes filtered out during the alignment step are now reported in the <cite>gtdbtk.bac120.summary.tsv</cite> or <cite>gtdbtk.ar53.summary.tsv</cite> as ‘Unclassified Bacteria/Archaea’</p></li>
<li><p><cite>–write_single_copy_genes</cite> flag in now available in the <cite>classify_wf</cite> and <cite>de_novo_wf</cite> workflows.</p></li>
</ul>
<p>Features:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/392">#392</a>) <cite>–write_single_copy_genes</cite> flag available in workflows.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/392">#387</a>) specific memory requirements set in classify_wf depending on the classification approach.</p></li>
</ul>
</div>
<div class="section" id="id54">
<h2>2.0.0<a class="headerlink" href="#id54" title="Permalink to this headline">¶</a></h2>
<p>Major changes:</p>
<ul class="simple">
<li><p>GTDB-TK now uses a <strong>divide-and-conquer</strong> approach where the bacterial reference tree is split into multiple order-level subtrees. This reduces the memory requirements of GTDB-Tk from <strong>320 GB</strong> of RAM when using the full GTDB R07-RS207 reference tree to approximately <strong>35 GB</strong>. A manuscript describing this approach is in preparation. If you wish to continue using the full GTDB reference tree use the <cite>–full-tree</cite> flag.</p></li>
<li><p>Archaeal classification now uses a refined set of 53 archaeal-specific marker genes based on the recent publication by <a class="reference external" href="https://www.nature.com/articles/s41467-020-17408-w">Dombrowski et al., 2020</a>. This set of archaeal marker genes is now used by GTDB for curating the archaeal taxonomy.</p></li>
<li><p>By default, all directories containing intermediate results are <strong>now removed</strong> by default at the end of the <cite>classify_wf</cite> and <cite>de_novo_wf</cite> pipelines. If you wish to retain these intermediates files use the <cite>–keep-intermediates</cite> flag.</p></li>
<li><p>All MSA files produced by the <cite>align</cite> step are now compressed with gzip.</p></li>
<li><p>The classification summary and failed genomes files are now the only files linked in the root directory of <cite>classify_wf</cite>.</p></li>
</ul>
<p>Features:</p>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/373">#373</a>) <cite>convert_to_itol</cite> to convert trees into iTOL format</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/369">#369</a>) Output FASTA files are compressed by default</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/369">#369</a>) Intermediate files will be removed by default when using classify/de-novo workflows unless specified by <cite>–keep_intermediates</cite></p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/362">#362</a>) Add –genes flag for Error</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/360">#360</a> / <a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/356">#356</a>) A warning will be displayed if pplacer fails to place a genome</p></li>
</ul>
<p><strong>Important</strong></p>
<ul class="simple">
<li><p>This version is <strong>not</strong> backwards compatible with GTDB release 202.</p></li>
<li><p>This version requires a <a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release207/207.0/auxillary_files/gtdbtk_r207_data.tar.gz">new reference package</a></p></li>
</ul>
</div>
<div class="section" id="id61">
<h2>1.7.0<a class="headerlink" href="#id61" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/336">#336</a>) Warn the user if they have provided an incorrectly formatted taxonomy file.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/348">#348</a>) Gracefully exit the program if no single copy hits could be identified.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/351">#351</a>) Fixed an issue where GTDB-Tk would crash if spaces were present in the reference data path.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/354">#354</a>) Added optional <code class="docutils literal notranslate"><span class="pre">--tmpdir</span></code> argument to set temporary directory (thanks <a class="reference external" href="https://github.com/tr11-sanger">tr11-sanger</a>!).</p></li>
</ul>
</div>
<div class="section" id="id66">
<h2>1.6.0<a class="headerlink" href="#id66" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/337">#337</a>) Set minimum <cite>tqdm</cite> version to <cite>4.35.0</cite></p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/335">#335</a>) Fixed typo in output log messages (&#64;fplaza)</p></li>
<li><p>Removed the option to re-calculate RED values (<cite>–recalculate_red</cite>)</p></li>
</ul>
</div>
<div class="section" id="id69">
<h2>1.5.1<a class="headerlink" href="#id69" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/327">#327</a>) Disallow spaces in genome names/file paths due to downstream application issues.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/326">#326</a>) Disallow genome names that are blank.</p></li>
</ul>
</div>
<div class="section" id="id72">
<h2>1.5.0<a class="headerlink" href="#id72" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/311">#311</a>) Updated GTDB-Tk to support R202.
See <a class="reference external" href="https://ecogenomics.github.io/GTDBTk/installing/index.html#gtdb-tk-reference-data">https://ecogenomics.github.io/GTDBTk/installing/index.html#gtdb-tk-reference-data</a> for instructions on downloading R202.</p></li>
</ul>
</div>
<div class="section" id="id74">
<h2>1.4.2<a class="headerlink" href="#id74" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/311">#311</a>) Fixed –scratch_dir not working in v 1.4.1 for classify_wf</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/311">#312</a>) Automatic drop of genome leads to error in downstream modules of classify_wf</p></li>
</ul>
</div>
<div class="section" id="id77">
<h2>1.4.1<a class="headerlink" href="#id77" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Updated GitHub CI/CD to trigger docker build / tag version on release.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/255">#255</a>) (<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/297">#297</a>)
Fixed <code class="docutils literal notranslate"><span class="pre">'Namespace'</span> <span class="pre">object</span> <span class="pre">has</span> <span class="pre">no</span> <span class="pre">attribute</span></code> errors by adding default arguments to argparse.</p></li>
</ul>
</div>
<div class="section" id="id80">
<h2>1.4.0<a class="headerlink" href="#id80" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Check if stdout is being piped to a file before adding colour.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/283">#283</a>) Significantly improved <code class="docutils literal notranslate"><span class="pre">classify</span></code> performance (noticeable when running trees &gt; 1,000 taxa).</p></li>
<li><p>Automatically cap pplacer CPUs to 64 unless specifying <code class="docutils literal notranslate"><span class="pre">--pplacer_cpus</span></code> to prevent pplacer from hanging.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/262">#262</a>) Added <code class="docutils literal notranslate"><span class="pre">--write_single_copy_genes</span></code> to the <code class="docutils literal notranslate"><span class="pre">identify</span></code> command. Writes unaligned single-copy AR53/BAC120 marker genes to disk.</p></li>
<li><p>When running <code class="docutils literal notranslate"><span class="pre">-version</span></code> warn if GTDB-Tk is not running the most up-to-date version (disable via <code class="docutils literal notranslate"><span class="pre">GTDBTK_VER_CHECK</span> <span class="pre">=</span> <span class="pre">False</span></code> in <code class="docutils literal notranslate"><span class="pre">config.py</span></code>). If GTDB-Tk encounters an error it will silently continue (3 second timeout).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/276">#276</a>) Renamed the column <code class="docutils literal notranslate"><span class="pre">aa_percent</span></code> to <code class="docutils literal notranslate"><span class="pre">msa_percent</span></code> in <code class="docutils literal notranslate"><span class="pre">summary.tsv</span></code> (produced by <code class="docutils literal notranslate"><span class="pre">classify</span></code>).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/286">#286</a>) Fixed a file not found error when the reference data is a symbolic link (thanks <a class="reference external" href="https://github.com/davidealbanese">davidealbanese</a>!).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/277">#277</a>) Fixed an issue where if the user overrides the translation table using the optional 3rd column in the batchfile, the other coding density would appear as -100. Both translation table densities are now reported.</p></li>
<li><p>The <a class="reference internal" href="commands/check_install.html#commands-check-install"><span class="std std-ref">check_install</span></a> command now also checks that all third party binaries can be found on the system path.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">align</span></code> step is now approximately 10x faster.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/289">#289</a>) Added <code class="docutils literal notranslate"><span class="pre">--min_af</span></code> to <code class="docutils literal notranslate"><span class="pre">classify</span></code> and <code class="docutils literal notranslate"><span class="pre">classify_wf</span></code> which allows the user to specify the minimum alignment fraction for FastANI.</p></li>
<li><p>Added the <code class="docutils literal notranslate"><span class="pre">--mash_db</span></code> command to re-use the GTDB-Tk Mash reference database in <code class="docutils literal notranslate"><span class="pre">ani_rep</span></code>.</p></li>
</ul>
</div>
<div class="section" id="id87">
<h2>1.3.0<a class="headerlink" href="#id87" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>This version of GTDB-Tk requires a new version of the GTDB-Tk reference package
(gtdbtk_r95_data.tar.gz) <a class="reference external" href="https://data.ace.uq.edu.au/public/gtdb/data/releases/release95/95.0/auxillary_files/gtdbtk_r95_data.tar.gz">available here</a>.</p></li>
<li><p>Updated reference package to use the <strong>GTDB Release 95</strong> taxonomy.</p></li>
<li><p>Report if the species-specific ANI circumscription criteria is satisfied in the <code class="docutils literal notranslate"><span class="pre">ani_closest.tsv</span></code> file output by <code class="docutils literal notranslate"><span class="pre">ani_rep</span></code>.</p></li>
<li><p>Estimated time until completion has been dampened.</p></li>
</ul>
</div>
<div class="section" id="id88">
<h2>1.2.0<a class="headerlink" href="#id88" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/241">#241</a>) Moved GTDB-Tk entry point to <code class="docutils literal notranslate"><span class="pre">__main__.py</span></code> instead of
<code class="docutils literal notranslate"><span class="pre">bin/gtdbtk</span></code> to support execution in some HPC systems (<code class="docutils literal notranslate"><span class="pre">gtdbtk</span></code> will still be aliased on install).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/251">#251</a>) Allow parsing of FastANI v1.0 output files. However, a warning will be displayed to update FastANI.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/254">#254</a>) Fixed an issue where <code class="docutils literal notranslate"><span class="pre">--scratch_dir</span></code> would fail, and not clean-up the mmap file.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/242">#242</a>) Added the <code class="docutils literal notranslate"><span class="pre">decorate</span></code> command allowing the <code class="docutils literal notranslate"><span class="pre">de</span> <span class="pre">novo</span> <span class="pre">workflow</span></code> to be run</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/244">#244</a>) Added the <code class="docutils literal notranslate"><span class="pre">infer_rank</span></code> method which established the taxonomic ranks of internal nodes of user trees based on RED</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/248">#248</a>) If the identify command is run on the same directory, genomes which were already processed will be skipped.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/248">#248</a>) Improved <code class="docutils literal notranslate"><span class="pre">pplacer</span></code> output with running the <code class="docutils literal notranslate"><span class="pre">classify</span></code> command.</p></li>
</ul>
</div>
<div class="section" id="id96">
<h2>1.1.0<a class="headerlink" href="#id96" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>In rare cases pplacer would assign an empty taxonomy string which would raise an error.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/229">#229</a>) Genomes using windows line carriage <code class="docutils literal notranslate"><span class="pre">\r\n</span></code> would raise an error.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/227">#227</a>) CentOS machines would fail when using <code class="docutils literal notranslate"><span class="pre">~</span></code> in paths.</p></li>
<li><p>The bac120 symlink was pointing to the archaeal tree when using the <code class="docutils literal notranslate"><span class="pre">root</span></code> command.</p></li>
<li><p>Updated the <code class="docutils literal notranslate"><span class="pre">gtdb_to_ncbi_majority_vote.py</span></code> script for translating taxonomy.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/195">#195</a>) Added the <code class="docutils literal notranslate"><span class="pre">--pplacer_cpus</span></code> argument to specify the number of pplacer threads when running <code class="docutils literal notranslate"><span class="pre">classify</span></code> and <code class="docutils literal notranslate"><span class="pre">classify_wf</span></code> (#195).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/198">#198</a>) The <code class="docutils literal notranslate"><span class="pre">--debug</span></code> flag of <code class="docutils literal notranslate"><span class="pre">align</span></code> outputs aligned markers to disk before trimming.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/225">#225</a>) An optional third column in the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> will specify an override to which translation table should be used.
Leave blank to automatically determine the translation table (default).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/131">#131</a>) Users can now specify genomes which have NCBI accessions, as long as they are not GTDB-Tk
representatives (a warning will be raised).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/191">#191</a>) Added a new command <code class="docutils literal notranslate"><span class="pre">ani_rep</span></code> which calculates the ANI of input genomes to all GTDB
representative genomes.</p></li>
<li><p>This command uses <a class="reference external" href="https://github.com/marbl/Mash">Mash</a> in a pre-filtering step. If pre-filtering is enabled (default)
then <code class="docutils literal notranslate"><span class="pre">mash</span></code> will need to be on the system path. To disable pre-filtering use the <code class="docutils literal notranslate"><span class="pre">--no_mash</span></code> flag.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/235">#230</a>) Improved how markers are used in determining the correct domain, and gene selection for the alignment.</p></li>
</ul>
</div>
<div class="section" id="id105">
<h2>1.0.2<a class="headerlink" href="#id105" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Fixed an issue where FastANI threads would timeout with <code class="docutils literal notranslate"><span class="pre">FastANI</span> <span class="pre">returned</span> <span class="pre">a</span> <span class="pre">non-zero</span> <span class="pre">exit</span> <span class="pre">code.</span></code></p></li>
<li><p>Versions affected: <code class="docutils literal notranslate"><span class="pre">1.0.0</span></code>, and <code class="docutils literal notranslate"><span class="pre">1.0.1</span></code>.</p></li>
</ul>
</div>
<div class="section" id="id106">
<h2>1.0.0<a class="headerlink" href="#id106" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Migrated to <strong>Python 3</strong>, you must be running at least <strong>Python 3.6</strong> or later to use this version.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">check_install</span></code> now does an exhaustive check of the reference data.</p></li>
<li><p>Resolved an issue where gene calling would fail for low quality genomes (#192).</p></li>
<li><p>Improved FastANI multiprocessing performance.</p></li>
<li><p>Third party software versions are reported where possible.</p></li>
</ul>
</div>
<div class="section" id="id107">
<h2>0.3.3<a class="headerlink" href="#id107" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>A bug has been fixed which affected <code class="docutils literal notranslate"><span class="pre">classify</span></code> and <code class="docutils literal notranslate"><span class="pre">classify_wf</span></code> when using the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>
argument with genome IDs that differed from the FASTA filename. This issue resulted in
the assigned taxonomy being derived only from tree placement without any ANI
calculations being considered. Consequently, in some cases genomes may have been classified as a new
species within a genus when they should have been assigned to an existing species. If you have genomes
with species assignments this bug did not impact you.</p></li>
<li><p>Progress is now displayed for: hmmalign, and pplacer.</p></li>
<li><p>Fixed an issue where the <code class="docutils literal notranslate"><span class="pre">root</span></code> command could not be run independently.</p></li>
<li><p>Improved MSA masking performance.</p></li>
</ul>
</div>
<div class="section" id="id108">
<h2>0.3.2<a class="headerlink" href="#id108" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>FastANI calculations are more robust.</p></li>
<li><p>Optimisation of RED calculations.</p></li>
<li><p>Improved output messages when errors are encountered.</p></li>
</ul>
</div>
<div class="section" id="id109">
<h2>0.3.1<a class="headerlink" href="#id109" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Pplacer taxonomy is now available in the summary file.</p></li>
<li><p>FastANI species assignment will be selected over phylogenetic placement (Topology case).</p></li>
</ul>
</div>
<div class="section" id="id110">
<h2>0.3.0<a class="headerlink" href="#id110" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Best translation table displayed in summary file.</p></li>
<li><p>GTDB-Tk now supports gzipped genomes as inputs (<code class="docutils literal notranslate"><span class="pre">--extension</span> <span class="pre">gz</span></code>).</p></li>
<li><p>By default, GTDB-Tk uses precalculated RED values.</p></li>
<li><p>New option to recalculate RED value during classify step (<code class="docutils literal notranslate"><span class="pre">--recalculate_red</span></code>).</p></li>
<li><p>New option to export the untrimmed reference MSA files.</p></li>
<li><p>New option to skip_trimming during align step.</p></li>
<li><p>New option to use a custom taxonomy file when rooting a tree.</p></li>
<li><p>New FAQ page available.</p></li>
<li><p>New output structure.</p></li>
</ul>
</div>
<div class="section" id="id111">
<h2>0.2.1<a class="headerlink" href="#id111" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Species classification is now based strictly on the ANI to reference genomes</p></li>
<li><p>The “classify” function now reports the closest reference genome in the summary file even if the ANI is &lt;95%</p></li>
<li><p>The summary.tsv file has 4 new columns: aa_percent, red_values, fastani_reference_radius, and warnings</p></li>
<li><p>By default, the “align” function now performs the same MSA trimming used by the GTDB</p></li>
<li><p>New pplacer support for writing to a scratch file (<code class="docutils literal notranslate"><span class="pre">--mmap-file</span></code> option)</p></li>
<li><p>Random seed option for MSA trimming has been added to allow for reproducible results</p></li>
<li><p>Configuration of the data directory is now set using the environment variable <code class="docutils literal notranslate"><span class="pre">GTDBTK_DATA_PATH</span></code> (see pip installation)</p></li>
<li><p>Perl dependencies has been removed</p></li>
<li><p>Python libraries biolib, mpld3 and jinja have been removed</p></li>
<li><p>This version requires a new version of the GTDB-Tk data package (gtdbtk.r86_v2_data.tar.gz) available <a class="reference external" href="https://data.ace.uq.edu.au/public/gtdbtk/release_86/">here</a></p></li>
</ul>
</div>
<div class="section" id="id112">
<h2>0.1.3<a class="headerlink" href="#id112" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>GTDB-Tk v0.1.3 has been released and addresses an issue with species assignments based on the placement of
genomes in the reference tree. This impacted species assignment when submitting multiple closely related genomes.
Species assignments reported by ANI were not impacted.</p></li>
</ul>
</div>
<div class="section" id="id113">
<h2>0.1.0<a class="headerlink" href="#id113" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Updated to R86, requires <a class="reference external" href="https://data.ace.uq.edu.au/public/gtdbtk/release_86/">release 86</a> to run.</p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="references.html" class="btn btn-neutral float-right" title="References" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="examples/classify_wf.html" class="btn btn-neutral float-left" title="Example" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2025, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>