

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>FAQ &mdash; GTDB-Tk 2.5.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/faq.html" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="_static/custom.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Performance and Accuracy" href="performance/index.html" />
    <link rel="prev" title="Docker" href="installing/docker.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">FAQ</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#taxonomy-faq">Taxonomy FAQ</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#why-is-there-a-discrepancy-in-the-naming-system-between-gtdb-tk-and-ncbi-or-silva-taxonomic-names">Why is there a discrepancy in the naming system between GTDB-Tk and NCBI or Silva taxonomic names?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#why-do-i-see-discrepancies-in-classification-for-closely-related-genomes">Why do I see discrepancies in classification for closely related genomes?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#can-you-combine-the-bacterial-and-archaeal-trees-into-a-single-tree">Can you combine the bacterial and archaeal trees into a single tree?</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#gtdb-tk-faq">GTDB-Tk FAQ</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#gtdb-tk-reaches-the-memory-limit-pplacer-crashes">GTDB-Tk reaches the memory limit / pplacer crashes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#how-is-gtdb-tk-validating-species-assignments-using-average-nucleotide-identity">How is GTDB-Tk validating species assignments using average nucleotide identity?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#what-is-the-difference-between-the-mutually-exclusive-options-mash-db-and-skip-ani-screen">What is the difference between the mutually exclusive options <code class="docutils literal notranslate"><span class="pre">--mash_db</span></code> and <code class="docutils literal notranslate"><span class="pre">--skip_ani_screen</span></code>?</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#deprecated-faq">Deprecated FAQ</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#why-is-fastani-using-more-threads-than-allocated">Why is FastANI using more threads than allocated?</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1"><a class="reference internal" href="commands/index.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>FAQ</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/faq.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="faq">
<span id="id1"></span><h1>FAQ<a class="headerlink" href="#faq" title="Permalink to this headline">¶</a></h1>
<div class="section" id="taxonomy-faq">
<h2>Taxonomy FAQ<a class="headerlink" href="#taxonomy-faq" title="Permalink to this headline">¶</a></h2>
<div class="section" id="why-is-there-a-discrepancy-in-the-naming-system-between-gtdb-tk-and-ncbi-or-silva-taxonomic-names">
<h3>Why is there a discrepancy in the naming system between GTDB-Tk and NCBI or Silva taxonomic names?<a class="headerlink" href="#why-is-there-a-discrepancy-in-the-naming-system-between-gtdb-tk-and-ncbi-or-silva-taxonomic-names" title="Permalink to this headline">¶</a></h3>
<p>GTDB-Tk uses the GTDB taxonomy (<a class="reference external" href="https://gtdb.ecogenomic.org/">https://gtdb.ecogenomic.org/</a>).
This taxonomy is similar, but not identical to NCBI and Silva.
In many cases the GTDB taxonomy more strictly follows the nomenclatural rules for rank suffixes which is why there is Nitrospirota instead of Nitrospirae.</p>
</div>
<div class="section" id="why-do-i-see-discrepancies-in-classification-for-closely-related-genomes">
<h3>Why do I see discrepancies in classification for closely related genomes?<a class="headerlink" href="#why-do-i-see-discrepancies-in-classification-for-closely-related-genomes" title="Permalink to this headline">¶</a></h3>
<p>Discrepancies in taxonomic assignments can occur when working with closely related genomes. GTDB-Tk uses both the <strong>Relative Evolutionary Divergence (RED)</strong> value and the <strong>placement of the genome in the reference tree</strong> to determine the best taxonomic classification.
In most cases taxonomic assignments are robust, but it is possible for highly similar genomes  to have sufficiently different protein sequences that their placements in the reference tree will vary slightly. This can lead to:</p>
<ol class="arabic simple">
<li><p>placement on different but closely related branches, or</p></li>
<li><p>placement on the same branch but at different depths.</p></li>
</ol>
<p>This uncertainty in the placement of even closely related genomes can result in genomes being erroneously assigned to different taxa (e.g. closely related sister families).</p>
<p><strong>What should I do?</strong></p>
<p>If you run into this situation, here are a few strategies you can use:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>Dereplicate your genomes.</dt><dd><p>If you have a cluster of closely related genomes representing a single species, consider dereplicating them to pick a single representative based on genome quality metrics (for example, using a tool like <a class="reference external" href="https://drep.readthedocs.io/en/latest/index.html">dRep</a> or <a class="reference external" href="https://github.com/wwood/galah">Galah</a>). The taxonomic assignment for this species representative genomes can then be propagated to the other genomes in the cluster.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Build a de novo tree</dt><dd><p>Another option is to construct a de novo tree including your closely related genomes plus a suitable outgroup. For instance, if you have three genomes classified as:</p>
<ul class="simple">
<li><p><em>c__Atribacteria;o__Atribacterales;f__Atribacteraceae</em></p></li>
<li><p><em>c__Atribacteria;o__Atribacterales;f__Thermatribacteraceae</em></p></li>
<li><p><em>c__Atribacteria;o__Atribacterales;f__</em></p></li>
</ul>
</dd>
</dl>
</li>
</ol>
<p>you could generate a tree for the class Atribacteria (with one outgroup class) and examine whether your genomes cluster together despite their different family-level assignments.</p>
</div>
<div class="section" id="can-you-combine-the-bacterial-and-archaeal-trees-into-a-single-tree">
<h3>Can you combine the bacterial and archaeal trees into a single tree?<a class="headerlink" href="#can-you-combine-the-bacterial-and-archaeal-trees-into-a-single-tree" title="Permalink to this headline">¶</a></h3>
<p>The bacterial and archaeal trees are inferred from different marker genes. Currently, the correct rootings of these trees remain an open area of research.
GTDB-Tk does not provide a tool to merge the trees but It is possible to artificially combine them by manipulating the Newick files.
One solution would be to use (<a class="reference external" href="https://dendropy.org/">DendroPy</a>); a Python library used for phylogenetic computing.</p>
</div>
</div>
<div class="section" id="gtdb-tk-faq">
<h2>GTDB-Tk FAQ<a class="headerlink" href="#gtdb-tk-faq" title="Permalink to this headline">¶</a></h2>
<div class="section" id="gtdb-tk-reaches-the-memory-limit-pplacer-crashes">
<span id="faq-pplacer"></span><h3>GTDB-Tk reaches the memory limit / pplacer crashes<a class="headerlink" href="#gtdb-tk-reaches-the-memory-limit-pplacer-crashes" title="Permalink to this headline">¶</a></h3>
<p>The host may report that GTDB-Tk has exceeded the memory requirements due to how <code class="docutils literal notranslate"><span class="pre">pplacer</span></code> is implemented.
Briefly, this is only the reported value and is not true for how much memory is actually in use.</p>
<p>When pplacer runs, it goes through several steps notable detailed below:</p>
<ol class="arabic">
<li><p>pplacer requests an uninitialised chunk of memory from the host (say, 150 GB).</p>
<ul class="simple">
<li><p>VIRT = 150 GB <code class="docutils literal notranslate"><span class="pre">[virtual</span> <span class="pre">memory</span> <span class="pre">memory,</span> <span class="pre">i.e.</span> <span class="pre">the</span> <span class="pre">amount</span> <span class="pre">of</span> <span class="pre">memory</span> <span class="pre">mapped]</span></code></p></li>
<li><p>RES ~= 0 GB <code class="docutils literal notranslate"><span class="pre">[amount</span> <span class="pre">of</span> <span class="pre">memory</span> <span class="pre">physically</span> <span class="pre">in</span> <span class="pre">use]</span></code></p></li>
</ul>
</li>
<li><p>pplacer starts caching information to that chunk of memory</p>
<ul class="simple">
<li><p>VIRT = 150 GB</p></li>
<li><p>RES -&gt; 150 GB  (slowly increases to 150 GB as caching information is written)</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If pplacer crashes here, you likely don’t have enough free memory on the server.</p>
</div>
</li>
<li><p>pplacer starts placing genomes into the tree</p>
<ol class="arabic">
<li><p>the main pplacer thread forks for each <code class="docutils literal notranslate"><span class="pre">--pplacer_cpus</span></code> or <code class="docutils literal notranslate"><span class="pre">--cpus</span></code> (if not specified).</p>
<ul class="simple">
<li><p>Unix uses the <a class="reference external" href="https://en.wikipedia.org/wiki/Copy-on-write">copy-on-write</a> method for each child.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This means that a new thread will appear using the same amount of memory as the parent (150 GB).
Due to how copy-on-write is implemented, this is the same memory space and is not using any additional physical memory.</p>
<p>Therefore, the host will report pplacer using a total of <code class="docutils literal notranslate"><span class="pre">PARENT_MEMORY</span> <span class="pre">*</span> <span class="pre">(N_CHILDREN</span> <span class="pre">+</span> <span class="pre">1)</span></code> GB.</p>
</div>
</li>
<li><p>each worker will only read from the memory space and exit once the queue of query genomes is depleted.</p></li>
</ol>
</li>
</ol>
<p>For example, running GTDB-Tk with on the bacterial tree (requires 150 GB of memory) with 1 CPU will require 150 GB of physical
memory, but the host will report 300 GB of memory in use.</p>
<p>Using the <code class="docutils literal notranslate"><span class="pre">--scratch_dir</span></code> parameter and <code class="docutils literal notranslate"><span class="pre">--pplacer_cpus</span> <span class="pre">1</span></code> may help.</p>
</div>
<div class="section" id="how-is-gtdb-tk-validating-species-assignments-using-average-nucleotide-identity">
<h3>How is GTDB-Tk validating species assignments using average nucleotide identity?<a class="headerlink" href="#how-is-gtdb-tk-validating-species-assignments-using-average-nucleotide-identity" title="Permalink to this headline">¶</a></h3>
<p>GTDB-Tk uses <a class="reference external" href="https://github.com/bluenote-1577/skani">skani</a> ( it was using fastANI until v2.3.2) to estimate the ANI between genomes.
A query genome is only classified as belonging to the same species as a reference genome if the ANI between the
genomes is within the species ANI circumscription radius (typically, 95%) and the alignment fraction (AF) is &gt;=0.5.
In some circumstances, the phylogenetic placement of a query genome may not support the species assignment.
GTDB r207+ strictly uses ANI to circumscribe species and GTDB-Tk follows this methodology.
The species-specific ANI circumscription radii are available from the <a class="reference external" href="https://gtdb.ecogenomic.org/">GTDB</a> website.</p>
</div>
<div class="section" id="what-is-the-difference-between-the-mutually-exclusive-options-mash-db-and-skip-ani-screen">
<h3>What is the difference between the mutually exclusive options <code class="docutils literal notranslate"><span class="pre">--mash_db</span></code> and <code class="docutils literal notranslate"><span class="pre">--skip_ani_screen</span></code>?<a class="headerlink" href="#what-is-the-difference-between-the-mutually-exclusive-options-mash-db-and-skip-ani-screen" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">Starting with GTDB-Tk v2.2+, the <code class="docutils literal notranslate"><span class="pre">classify_wf</span></code> and <code class="docutils literal notranslate"><span class="pre">classify</span></code> function require an extra parameter to run: <code class="docutils literal notranslate"><span class="pre">--mash_db</span></code> or <code class="docutils literal notranslate"><span class="pre">--skip_ani_screen</span></code>.</div>
<div class="line">With this new version of Tk, The first stage of <code class="docutils literal notranslate"><span class="pre">classify</span></code> pipelines (<code class="docutils literal notranslate"><span class="pre">classify_wf</span></code> and <code class="docutils literal notranslate"><span class="pre">classify</span></code>) is to compare all user genomes to all reference genomes and annotate them, if possible, based on ANI matches.</div>
<div class="line">Using the <code class="docutils literal notranslate"><span class="pre">--mash_db</span></code> option will indicate to GTDB-Tk the path of the sketched Mash database require for ANI screening.</div>
<div class="line">If no database are available ( i.e. this is the first time running classify ), the <code class="docutils literal notranslate"><span class="pre">--mash_db</span></code> option will sketch a new Mash database that can be used for subsequent calls.</div>
<div class="line">The <code class="docutils literal notranslate"><span class="pre">--skip_ani_screen</span></code> option will skip the pre-screening step and classify all genomes similar to previous versions of GTDB-Tk.</div>
</div>
</div>
</div>
<div class="section" id="deprecated-faq">
<h2>Deprecated FAQ<a class="headerlink" href="#deprecated-faq" title="Permalink to this headline">¶</a></h2>
<div class="section" id="why-is-fastani-using-more-threads-than-allocated">
<h3>Why is FastANI using more threads than allocated?<a class="headerlink" href="#why-is-fastani-using-more-threads-than-allocated" title="Permalink to this headline">¶</a></h3>
<p>If you are using FastANI version 1.33 then you may run into an issue where FastANI will use more threads than you allocate.
This can be problematic if running GTDB-Tk on a HPC where you have a limited number of threads available.</p>
<p>This issue has been <a class="reference external" href="https://github.com/ParBLiSS/FastANI/issues/101">reported to the FastANI developers (#101)</a>.</p>
<p>Depending on how you installed GTDB-Tk there are different ways to downgrade FastANI to version 1.32.</p>
<p><strong>Manual:</strong></p>
<p>Simplify download and install the FastANI binary from <a class="reference external" href="https://github.com/ParBLiSS/FastANI/releases/tag/v1.32">here</a>.</p>
<p><strong>Conda:</strong></p>
<p>From GTDB-Tk v2.0.0 the conda environment will automatically have FastANI v1.3 installed. Otherwise run:</p>
<p><code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">bioconda</span> <span class="pre">fastani==1.32</span></code></p>
<p><strong>Docker:</strong></p>
<p>From GTDB-Tk v2.2.2 the Docker container will automatically have FastANI v1.32 installed. Otherwise, manually
build the container from the <a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/blob/master/Dockerfile">Dockerfile</a>, making
sure to specify FastANI v1.32.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="performance/index.html" class="btn btn-neutral float-right" title="Performance and Accuracy" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="installing/docker.html" class="btn btn-neutral float-left" title="Docker" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2025, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>