

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>References &mdash; GTDB-Tk 1.4.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/custom.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Change log" href="changelog.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.4.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
</ul>
<p class="caption"><span class="caption-text">Running GTDB-Tk</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="commands/index.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption"><span class="caption-text">About</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Change log</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>References</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/references.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="references">
<span id="id1"></span><h1>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<p><strong>GTDB-Tk is described in:</strong></p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 100%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Chaumeil PA, et al. 2019. <a class="reference external" href="https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btz848/5626182">GTDB-Tk: A toolkit to classify genomes with the Genome Taxonomy Database</a>. <em>Bioinformatics</em>, btz848.</p></td>
</tr>
</tbody>
</table>
<p><strong>The Genome Taxonomy Database (GTDB) is described in:</strong></p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 100%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Parks DH, et al. 2020. <a class="reference external" href="https://rdcu.be/b3OI7">A complete domain-to-species taxonomy for Bacteria and Archaea</a>. <em>Nature Biotechnology</em>, <a class="reference external" href="https://doi.org/10.1038/s41587-020-0501-8">https://doi.org/10.1038/s41587-020-0501-8</a>.</p></td>
</tr>
<tr class="row-even"><td><p>Parks DH, et al. 2018. <a class="reference external" href="https://www.nature.com/articles/nbt.4229">A standardized bacterial taxonomy based on genome phylogeny substantially revises the tree of life</a>. <em>Nature Biotechnology</em>, <a class="reference external" href="http://dx.doi.org/10.1038/nbt.4229">http://dx.doi.org/10.1038/nbt.4229</a>.</p></td>
</tr>
</tbody>
</table>
<p><strong>We strongly encourage you to cite the following 3rd party dependencies:</strong></p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Software</p></th>
<th class="head"><p>Reference</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="http://compbio.ornl.gov/prodigal/">Prodigal</a></p></td>
<td><p>Hyatt D, et al. 2010. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/20211023">Prodigal: prokaryotic gene recognition and translation initiation site identification</a>. <em>BMC Bioinformatics</em>, 11:119. doi: 10.1186/1471-2105-11-119.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="http://hmmer.org/">HMMER</a></p></td>
<td><p>Eddy SR. 2011. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/22039361">Accelerated profile HMM searches</a>. <em>PLOS Comp. Biol.</em>, 7:e1002195.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="http://matsen.fhcrc.org/pplacer/">pplacer</a></p></td>
<td><p>Matsen FA, et al. 2010. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/21034504">pplacer: linear time maximum-likelihood and Bayesian phylogenetic placement of sequences onto a fixed reference tree</a>. <em>BMC Bioinformatics</em>, 11:538.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://github.com/ParBLiSS/FastANI">FastANI</a></p></td>
<td><p>Jain C, et al. 2019. <a class="reference external" href="https://www.nature.com/articles/s41467-018-07641-9">High-throughput ANI Analysis of 90K Prokaryotic Genomes Reveals Clear Species Boundaries</a>. <em>Nat. Communications</em>, doi: 10.1038/s41467-018-07641-9.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a></p></td>
<td><p>Price MN, et al. 2010. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2835736/">FastTree 2 - Approximately Maximum-Likelihood Trees for Large Alignments</a>. <em>PLoS One</em>, 5, e9490.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://github.com/marbl/Mash">Mash</a></p></td>
<td><p>Ondov BD, et al. 2016. <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x">Mash: fast genome and metagenome distance estimation using MinHash</a>. <em>Genome Biol</em> 17, 132. doi: doi: 10.1186/s13059-016-0997-x.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://dendropy.org/">DendroPy</a></p></td>
<td><p>Sukumaran, J. and Mark T. Holder. 2010. DendroPy: A Python library for phylogenetic computing. Bioinformatics 26: 1569-1571.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://numpy.org/">NumPy</a></p></td>
<td><p>Harris, C.R., Millman, K.J., van der Walt, S.J. et al. Array programming with NumPy. Nature 585, 357–362 (2020). DOI: <a class="reference external" href="https://doi.org/10.1038/s41586-020-2649-2">0.1038/s41586-020-2649-2</a></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://github.com/tqdm/tqdm">tqdm</a></p></td>
<td><p>DOI: <a class="reference external" href="https://doi.org/10.5281/zenodo.595120">10.5281/zenodo.595120</a></p></td>
</tr>
</tbody>
</table>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="changelog.html" class="btn btn-neutral float-left" title="Change log" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>