

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>classify &mdash; GTDB-Tk 2.4.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/commands/classify.html" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="classify_wf" href="classify_wf.html" />
    <link rel="prev" title="check_install" href="check_install.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.4.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Commands</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="align.html">align</a></li>
<li class="toctree-l2"><a class="reference internal" href="ani_rep.html">ani_rep</a></li>
<li class="toctree-l2"><a class="reference internal" href="check_install.html">check_install</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">classify</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#mutually exclusive required arguments">mutually exclusive required arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#required named arguments">required named arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mutually exclusive required arguments_repeat1">mutually exclusive required arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#optional Mash arguments">optional Mash arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#files-output">Files output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#input">Input</a></li>
<li class="toctree-l4"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="classify_wf.html">classify_wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="convert_to_itol.html">convert_to_itol</a></li>
<li class="toctree-l2"><a class="reference internal" href="de_novo_wf.html">de_novo_wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="decorate.html">decorate</a></li>
<li class="toctree-l2"><a class="reference internal" href="export_msa.html">export_msa</a></li>
<li class="toctree-l2"><a class="reference internal" href="identify.html">identify</a></li>
<li class="toctree-l2"><a class="reference internal" href="infer.html">infer</a></li>
<li class="toctree-l2"><a class="reference internal" href="root.html">root</a></li>
<li class="toctree-l2"><a class="reference internal" href="test.html">test</a></li>
<li class="toctree-l2"><a class="reference internal" href="trim_msa.html">trim_msa</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">Commands</a> &raquo;</li>
        
      <li>classify</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/commands/classify.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="classify">
<span id="commands-classify"></span><h1>classify<a class="headerlink" href="#classify" title="Permalink to this headline">¶</a></h1>
<p>Determine taxonomic classification of genomes.</p>
<div class="section" id="arguments">
<h2>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">gtdbtk</span> <span class="n">classify</span> <span class="p">(</span><span class="o">--</span><span class="n">genome_dir</span> <span class="n">GENOME_DIR</span> <span class="o">|</span> <span class="o">--</span><span class="n">batchfile</span> <span class="n">BATCHFILE</span><span class="p">)</span>
                       <span class="o">--</span><span class="n">align_dir</span> <span class="n">ALIGN_DIR</span> <span class="o">--</span><span class="n">out_dir</span> <span class="n">OUT_DIR</span>
                       <span class="p">(</span><span class="o">--</span><span class="n">skip_ani_screen</span> <span class="o">|</span> <span class="o">--</span><span class="n">mash_db</span> <span class="n">MASH_DB</span><span class="p">)</span> <span class="p">[</span><span class="o">--</span><span class="n">no_mash</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">mash_k</span> <span class="n">MASH_K</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mash_s</span> <span class="n">MASH_S</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mash_v</span> <span class="n">MASH_V</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">mash_max_distance</span> <span class="n">MASH_MAX_DISTANCE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">EXTENSION</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">prefix</span> <span class="n">PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">cpus</span> <span class="n">CPUS</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">pplacer_cpus</span> <span class="n">PPLACER_CPUS</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">scratch_dir</span> <span class="n">SCRATCH_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">genes</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">min_af</span> <span class="n">MIN_AF</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tmpdir</span> <span class="n">TMPDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">debug</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="mutually exclusive required arguments">
<h3>mutually exclusive required arguments<a class="headerlink" href="#mutually exclusive required arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--genome_dir</kbd></dt>
<dd><p>directory containing genome files in FASTA format</p>
</dd>
<dt><kbd>--batchfile</kbd></dt>
<dd><p>path to file describing genomes - tab separated in 2 or 3 columns (FASTA file, genome ID, translation table [optional])</p>
</dd>
</dl>
</div>
<div class="section" id="required named arguments">
<h3>required named arguments<a class="headerlink" href="#required named arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--align_dir</kbd></dt>
<dd><p>output directory of ‘align’ command</p>
</dd>
<dt><kbd>--out_dir</kbd></dt>
<dd><p>directory to output files</p>
</dd>
</dl>
</div>
<div class="section" id="mutually exclusive required arguments_repeat1">
<h3>mutually exclusive required arguments<a class="headerlink" href="#mutually exclusive required arguments_repeat1" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--skip_ani_screen</kbd></dt>
<dd><p>Skip the ani_screening step to classify genomes using mash and skani.</p>
</dd>
<dt><kbd>--mash_db</kbd></dt>
<dd><p>path to save/read (if exists) the Mash reference sketch database (.msh)</p>
</dd>
</dl>
</div>
<div class="section" id="optional Mash arguments">
<h3>optional Mash arguments<a class="headerlink" href="#optional Mash arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--no_mash</kbd></dt>
<dd><p>skip pre-filtering of genomes using Mash</p>
</dd>
<dt><kbd>--mash_k</kbd></dt>
<dd><p>k-mer size [1-32]</p>
<p>Default: 16</p>
</dd>
<dt><kbd>--mash_s</kbd></dt>
<dd><p>maximum number of non-redundant hashes</p>
<p>Default: 5000</p>
</dd>
<dt><kbd>--mash_v</kbd></dt>
<dd><p>maximum p-value to keep [0-1]</p>
<p>Default: 1.0</p>
</dd>
<dt><kbd>--mash_max_distance</kbd></dt>
<dd><p>Maximum Mash distance to select a potential GTDB genome as representative of a user genome.</p>
<p>Default: 0.15</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>-x, --extension</kbd></dt>
<dd><p>extension of files to process, <code class="docutils literal notranslate"><span class="pre">gz</span></code> = gzipped</p>
<p>Default: “fna”</p>
</dd>
<dt><kbd>--prefix</kbd></dt>
<dd><p>prefix for all output files</p>
<p>Default: “gtdbtk”</p>
</dd>
<dt><kbd>--cpus</kbd></dt>
<dd><p>number of CPUs to use</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--pplacer_cpus</kbd></dt>
<dd><p>number of CPUs to use during pplacer placement</p>
</dd>
<dt><kbd>--scratch_dir</kbd></dt>
<dd><p>reduce pplacer memory usage by writing to disk (slower).</p>
</dd>
<dt><kbd>--genes</kbd></dt>
<dd><p>indicates input files contain predicted proteins as amino acids (skip gene calling).Warning: This flag will skip the ANI comparison steps (ani_screen and classification).</p>
</dd>
<dt><kbd>-f, --full_tree</kbd></dt>
<dd><p>use the unsplit bacterial tree for the classify step; this is the original GTDB-Tk approach (version &lt; 2) and requires more than 320 GB of RAM to load the reference tree</p>
</dd>
<dt><kbd>--min_af</kbd></dt>
<dd><p>minimum alignment fraction to assign genome to a species cluster</p>
<p>Default: 0.5</p>
</dd>
<dt><kbd>--tmpdir</kbd></dt>
<dd><p>specify alternative directory for temporary files</p>
<p>Default: “/tmp”</p>
</dd>
<dt><kbd>--debug</kbd></dt>
<dd><p>create intermediate files for debugging purposes</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="files-output">
<h2>Files output<a class="headerlink" href="#files-output" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>classify</dt><dd><ul>
<li><p><a class="reference internal" href="../files/summary.tsv.html#files-summary-tsv"><span class="std std-ref">[prefix].[domain].summary.tsv</span></a></p></li>
<li><p><a class="reference internal" href="../files/classify.tree.html#files-classify-tree"><span class="std std-ref">[prefix].backbone.[domain].classify.tree</span></a></p></li>
<li><p><a class="reference internal" href="../files/tree.mapping.tsv.html#files-tree-mapping-tsv"><span class="std std-ref">[prefix].[domain].tree.mapping.tsv</span></a></p></li>
<li><p><a class="reference internal" href="../files/classify.tree.html#files-classify-tree"><span class="std std-ref">[prefix].[domain].classify.tree.[index].tree</span></a></p></li>
<li><dl class="simple">
<dt>intermediate_results</dt><dd><ul>
<li><p><a class="reference internal" href="../files/classification_pplacer.tsv.html#files-classification-pplacer-tsv"><span class="std std-ref">[prefix].[domain].backbone.classification_pplacer.tsv</span></a></p></li>
<li><p><a class="reference internal" href="../files/classification_pplacer.tsv.html#files-classification-pplacer-tsv"><span class="std std-ref">[prefix].[domain].class_level.classification_pplacer_tree_[index].tsv</span></a></p></li>
<li><p><a class="reference internal" href="../files/msa.fasta.html#files-msa-fasta"><span class="std std-ref">[prefix].[domain].prescreened.msa.fasta</span></a></p></li>
<li><p><a class="reference internal" href="../files/red_dictionary.tsv.html#files-red-dictionary-tsv"><span class="std std-ref">[prefix].[domain].red_dictionary.tsv</span></a></p></li>
<li><dl class="simple">
<dt>pplacer</dt><dd><ul>
<li><p><a class="reference internal" href="../files/pplacer.domain.json.html#files-pplacer-domain-json"><span class="std std-ref">pplacer.backbone.[domain].json</span></a></p></li>
<li><p><a class="reference internal" href="../files/pplacer.domain.out.html#files-pplacer-domain-out"><span class="std std-ref">pplacer.backbone.[domain].out</span></a></p></li>
<li><dl class="simple">
<dt>tree_[index]</dt><dd><ul>
<li><p><a class="reference internal" href="../files/user_msa.fasta.html#files-user-msa-fasta"><span class="std std-ref">[prefix].[domain].user_msa.fasta</span></a></p></li>
<li><p><a class="reference internal" href="../files/pplacer.domain.out.html#files-pplacer-domain-out"><span class="std std-ref">pplacer.class_level.[domain].out</span></a></p></li>
<li><p><a class="reference internal" href="../files/pplacer.domain.json.html#files-pplacer-domain-json"><span class="std std-ref">pplacer.class_level.[domain].json</span></a></p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ani_screen</dt><dd><ul>
<li><dl class="simple">
<dt>intermediate_results</dt><dd><ul>
<li><dl class="simple">
<dt>mash</dt><dd><ul>
<li><p><a class="reference internal" href="../files/mash_distances.msh.html#files-mash-distances-msh"><span class="std std-ref">[prefix].mash_distances.tsv</span></a></p></li>
<li><p><a class="reference internal" href="../files/user_query_sketch.msh.html#files-user-query-sketch-msh"><span class="std std-ref">[prefix].user_query_sketch.msh</span></a></p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><p><a class="reference internal" href="../files/summary.tsv.html#files-summary-tsv"><span class="std std-ref">[prefix].[domain].summary.tsv</span></a></p></li>
<li><p><a class="reference internal" href="../files/gtdbtk.log.html#files-gtdbtk-log"><span class="std std-ref">[prefix].log</span></a></p></li>
<li><p><a class="reference internal" href="../files/gtdbtk.json.html#files-gtdbtk-json"><span class="std std-ref">[prefix].json</span></a></p></li>
<li><p><a class="reference internal" href="../files/gtdbtk.warnings.log.html#files-gtdbtk-warnings-log"><span class="std std-ref">[prefix].warnings.log</span></a></p></li>
</ul>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gtdbtk<span class="w"> </span>classify<span class="w"> </span>--align_dir<span class="w"> </span>align_3lines/<span class="w"> </span>--batchfile<span class="w"> </span>3lines_batchfile.tsv<span class="w"> </span>--out_dir<span class="w"> </span>3classify_ani<span class="w"> </span>--mash_db<span class="w"> </span>mash_db_dir/<span class="w"> </span>--cpus<span class="w"> </span><span class="m">20</span>
</pre></div>
</div>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>[2024-03-27 15:46:26] INFO: GTDB-Tk v2.3.2
[2024-03-27 15:46:26] INFO: gtdbtk classify --align_dir 500_align --out_dir 500_classify --mash_db mash_db.msh --cpus 90 --batchfile genomes/500_batchfile.tsv
[2024-03-27 15:46:26] INFO: Using GTDB-Tk reference data version r214: /srv/db/gtdbtk/official/release214_skani/release214
[2024-03-27 15:46:27] WARNING: Setting pplacer CPUs to 64, as pplacer is known to hang if &gt;64 are used. You can override this using: --pplacer_cpus
[2024-03-27 15:46:27] INFO: Loading reference genomes.
[2024-03-27 15:46:27] INFO: Using Mash version 2.2.2
[2024-03-27 15:46:27] INFO: Creating Mash sketch file: 500_classify/classify/ani_screen/intermediate_results/mash/gtdbtk.user_query_sketch.msh
[2024-03-27 15:46:29] INFO: Completed 500 genomes in 1.74 seconds (286.68 genomes/second).
[2024-03-27 15:46:29] INFO: Loading data from existing Mash sketch file: mash_db.msh
[2024-03-27 15:46:32] INFO: Calculating Mash distances.
[2024-03-27 15:47:17] INFO: Calculating ANI with skani v0.2.1.
[2024-03-27 15:47:28] INFO: Completed 4,383 comparisons in 10.51 seconds (417.14 comparisons/second).
[2024-03-27 15:47:30] INFO: 357 genome(s) have been classified using the ANI pre-screening step.
[2024-03-27 15:47:30] TASK: Placing 143 bacterial genomes into backbone reference tree with pplacer using 64 CPUs (be patient).
[2024-03-27 15:47:30] INFO: pplacer version: v1.1.alpha19-0-g807f6f3
[2024-03-27 15:50:19] INFO: Calculating RED values based on reference tree.
[2024-03-27 15:50:20] INFO: 143 out of 143 have an class assignments. Those genomes will be reclassified.
[2024-03-27 15:50:20] TASK: Placing 30 bacterial genomes into class-level reference tree 3 (1/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 15:58:09] INFO: Calculating RED values based on reference tree.
[2024-03-27 15:58:12] TASK: Traversing tree to determine classification method.
[2024-03-27 15:58:12] INFO: Completed 30 genomes in 0.02 seconds (1,364.18 genomes/second).
[2024-03-27 15:58:15] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 15:58:16] INFO: Completed 505 comparisons in 1.32 seconds (381.38 comparisons/second).
[2024-03-27 15:58:19] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 15:58:19] TASK: Placing 27 bacterial genomes into class-level reference tree 2 (2/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 16:05:53] INFO: Calculating RED values based on reference tree.
[2024-03-27 16:05:56] TASK: Traversing tree to determine classification method.
[2024-03-27 16:05:56] INFO: Completed 27 genomes in 0.04 seconds (606.99 genomes/second).
[2024-03-27 16:05:59] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 16:06:00] INFO: Completed 317 comparisons in 1.07 seconds (297.29 comparisons/second).
[2024-03-27 16:06:03] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 16:06:03] TASK: Placing 26 bacterial genomes into class-level reference tree 6 (3/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 16:12:28] INFO: Calculating RED values based on reference tree.
[2024-03-27 16:12:30] TASK: Traversing tree to determine classification method.
[2024-03-27 16:12:30] INFO: Completed 26 genomes in 0.05 seconds (497.16 genomes/second).
[2024-03-27 16:12:31] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 16:12:32] INFO: Completed 41 comparisons in 0.83 seconds (49.26 comparisons/second).
[2024-03-27 16:12:33] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 16:12:34] TASK: Placing 22 bacterial genomes into class-level reference tree 7 (4/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 16:18:24] INFO: Calculating RED values based on reference tree.
[2024-03-27 16:18:27] TASK: Traversing tree to determine classification method.
[2024-03-27 16:18:27] INFO: Completed 22 genomes in 0.03 seconds (715.55 genomes/second).
[2024-03-27 16:18:28] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 16:18:28] INFO: Completed 117 comparisons in 0.84 seconds (138.63 comparisons/second).
[2024-03-27 16:18:30] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 16:18:30] TASK: Placing 22 bacterial genomes into class-level reference tree 1 (5/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 16:26:01] INFO: Calculating RED values based on reference tree.
[2024-03-27 16:26:04] TASK: Traversing tree to determine classification method.
[2024-03-27 16:26:04] INFO: Completed 22 genomes in 0.05 seconds (486.20 genomes/second).
[2024-03-27 16:26:05] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 16:26:06] INFO: Completed 373 comparisons in 1.05 seconds (354.70 comparisons/second).
[2024-03-27 16:26:08] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 16:26:08] TASK: Placing 8 bacterial genomes into class-level reference tree 4 (6/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 16:32:45] INFO: Calculating RED values based on reference tree.
[2024-03-27 16:32:48] TASK: Traversing tree to determine classification method.
[2024-03-27 16:32:48] INFO: Completed 8 genomes in 0.15 seconds (52.49 genomes/second).
[2024-03-27 16:32:48] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 16:32:49] INFO: Completed 176 comparisons in 0.90 seconds (195.38 comparisons/second).
[2024-03-27 16:32:50] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 16:32:50] TASK: Placing 4 bacterial genomes into class-level reference tree 8 (7/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 16:35:30] INFO: Calculating RED values based on reference tree.
[2024-03-27 16:35:31] TASK: Traversing tree to determine classification method.
[2024-03-27 16:35:31] INFO: Completed 4 genomes in 0.00 seconds (5,959.93 genomes/second).
[2024-03-27 16:35:32] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 16:35:33] INFO: Completed 31 comparisons in 0.93 seconds (33.24 comparisons/second).
[2024-03-27 16:35:33] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 16:35:33] TASK: Placing 4 bacterial genomes into class-level reference tree 5 (8/8) with pplacer using 64 CPUs (be patient).
[2024-03-27 16:40:57] INFO: Calculating RED values based on reference tree.
[2024-03-27 16:40:59] TASK: Traversing tree to determine classification method.
[2024-03-27 16:40:59] INFO: Completed 4 genomes in 0.00 seconds (4,607.86 genomes/second).
[2024-03-27 16:40:59] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2024-03-27 16:41:00] INFO: Completed 46 comparisons in 0.86 seconds (53.34 comparisons/second).
[2024-03-27 16:41:00] INFO: 0 genome(s) have been classified using skani and pplacer.
[2024-03-27 16:41:00] WARNING: 5 of 500 genomes have a warning (see summary file).
[2024-03-27 16:41:00] INFO: Note that Tk classification mode is insufficient for publication of new taxonomic designations. New designations should be based on one or more de novo trees, an example of which can be produced by Tk in de novo mode.
[2024-03-27 16:41:00] INFO: Done.
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="classify_wf.html" class="btn btn-neutral float-right" title="classify_wf" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="check_install.html" class="btn btn-neutral float-left" title="check_install" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2024, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>