

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>classify_wf &mdash; GTDB-Tk 2.5.2 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/commands/classify_wf.html" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="convert_to_itol" href="convert_to_itol.html" />
    <link rel="prev" title="classify" href="classify.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.5.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Commands</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="align.html">align</a></li>
<li class="toctree-l2"><a class="reference internal" href="ani_rep.html">ani_rep</a></li>
<li class="toctree-l2"><a class="reference internal" href="check_install.html">check_install</a></li>
<li class="toctree-l2"><a class="reference internal" href="classify.html">classify</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">classify_wf</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#classify-workflow">Classify workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#mutually exclusive required arguments">mutually exclusive required arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#required named arguments">required named arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#input">Input</a></li>
<li class="toctree-l4"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="convert_to_itol.html">convert_to_itol</a></li>
<li class="toctree-l2"><a class="reference internal" href="de_novo_wf.html">de_novo_wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="decorate.html">decorate</a></li>
<li class="toctree-l2"><a class="reference internal" href="export_msa.html">export_msa</a></li>
<li class="toctree-l2"><a class="reference internal" href="identify.html">identify</a></li>
<li class="toctree-l2"><a class="reference internal" href="infer.html">infer</a></li>
<li class="toctree-l2"><a class="reference internal" href="root.html">root</a></li>
<li class="toctree-l2"><a class="reference internal" href="test.html">test</a></li>
<li class="toctree-l2"><a class="reference internal" href="trim_msa.html">trim_msa</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">Commands</a> &raquo;</li>
        
      <li>classify_wf</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/commands/classify_wf.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="classify-wf">
<span id="commands-classify-wf"></span><h1>classify_wf<a class="headerlink" href="#classify-wf" title="Permalink to this headline">¶</a></h1>
<div class="section" id="classify-workflow">
<h2>Classify workflow<a class="headerlink" href="#classify-workflow" title="Permalink to this headline">¶</a></h2>
<p>For arguments and output files, see each of the individual steps:</p>
<ul class="simple">
<li><p><a class="reference internal" href="identify.html#commands-identify"><span class="std std-ref">identify</span></a></p></li>
<li><p><a class="reference internal" href="align.html#commands-align"><span class="std std-ref">align</span></a></p></li>
<li><p><a class="reference internal" href="classify.html#commands-classify"><span class="std std-ref">classify</span></a></p></li>
</ul>
<p>The classify workflow consists of four steps: <code class="docutils literal notranslate"><span class="pre">ani_screen</span></code>, <code class="docutils literal notranslate"><span class="pre">identify</span></code>, <code class="docutils literal notranslate"><span class="pre">align</span></code>, and <code class="docutils literal notranslate"><span class="pre">classify</span></code>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">ani_screen</span></code> step compares user genomes against a <a class="reference external" href="https://www.nature.com/articles/s41592-023-02018-3">skani</a> database composed of all GTDB representative genomes.
User genomes classified with skani are not run through the rest of the pipeline (<code class="docutils literal notranslate"><span class="pre">identify</span></code>, <code class="docutils literal notranslate"><span class="pre">align</span></code>, <code class="docutils literal notranslate"><span class="pre">classify</span></code>)
and are reported in the summary file.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">identify</span></code> step calls genes using <a class="reference external" href="http://compbio.ornl.gov/prodigal/">Prodigal</a>,
and uses HMM models and the <a class="reference external" href="http://hmmer.org/">HMMER</a> package to identify the
120 bacterial and 53 archaeal marker genes used for phylogenetic inference
(<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/30148503">Parks et al., 2018</a>). Multiple
sequence alignments (MSA) are obtained by aligning marker genes to their respective HMM model.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">align</span></code> step concatenates the aligned marker genes and filters the concatenated MSA to approximately 5,000 amino acids.</p>
<p>Finally, the <code class="docutils literal notranslate"><span class="pre">classify</span></code> step uses <a class="reference external" href="http://matsen.fhcrc.org/pplacer/">pplacer</a> to find the maximum-likelihood
placement of each genome in the GTDB-Tk reference tree. GTDB-Tk classifies each genome based on its placement in the reference tree,
its relative evolutionary divergence, and/or average nucleotide identity (ANI) to reference genomes.</p>
<p>Results can be impacted by a lack of marker genes or contamination. We have validated GTDB-Tk on genomes
estimated to be ≥50% complete with ≤10% contamination consistent with community standards for medium or
higher quality single-amplified and metagenome-assembled genomes (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28787424">Bowers et al., 2017</a>).</p>
<p>The classify workflow can be run as follows:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gtdbtk<span class="w"> </span>classify_wf<span class="w"> </span>--genome_dir<span class="w"> </span>&lt;my_genomes&gt;<span class="w"> </span>--out_dir<span class="w"> </span>&lt;output_dir&gt;
</pre></div>
</div>
<p>This will process all genomes in the directory &lt;my_genomes&gt; using both bacterial and archaeal marker sets and place the results in &lt;output_dir&gt;. Genomes must be in FASTA format (gzip with the extension .gz is acceptable).
The location of genomes can also be specified using a batch file with the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> flag. The batch file is a two column file indicating the location of each genome and the desired genome identifier
(i.e., a Newick compatible alphanumeric string). These fields must be separated by a tab.</p>
<p>The workflow supports several optional flags, including:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">min_perc_aa</span></code>: allows filtering of genomes below a specified percentage of amino acids in the MSA</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cpus</span></code>: maximum number of CPUs to use</p></li>
</ul>
<p>The taxonomic classification of each bacterial and archaeal genome is contained in the
<a class="reference internal" href="../files/summary.tsv.html#files-summary-tsv"><span class="std std-ref">[prefix].[domain].summary.tsv</span></a>  output files.</p>
<p>For other flags please consult the command line interface.</p>
</div>
<div class="section" id="arguments">
<h2>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">gtdbtk</span> <span class="n">classify_wf</span> <span class="p">(</span><span class="o">--</span><span class="n">genome_dir</span> <span class="n">GENOME_DIR</span> <span class="o">|</span> <span class="o">--</span><span class="n">batchfile</span> <span class="n">BATCHFILE</span><span class="p">)</span>
                          <span class="o">--</span><span class="n">out_dir</span> <span class="n">OUT_DIR</span> <span class="p">[</span><span class="o">--</span><span class="n">skip_ani_screen</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">EXTENSION</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">min_perc_aa</span> <span class="n">MIN_PERC_AA</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">prefix</span> <span class="n">PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">genes</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">cpus</span> <span class="n">CPUS</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">pplacer_cpus</span> <span class="n">PPLACER_CPUS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">force</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">scratch_dir</span> <span class="n">SCRATCH_DIR</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">write_single_copy_genes</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep_intermediates</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">min_af</span> <span class="n">MIN_AF</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tmpdir</span> <span class="n">TMPDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">debug</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="mutually exclusive required arguments">
<h3>mutually exclusive required arguments<a class="headerlink" href="#mutually exclusive required arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--genome_dir</kbd></dt>
<dd><p>directory containing genome files in FASTA format</p>
</dd>
<dt><kbd>--batchfile</kbd></dt>
<dd><p>path to file describing genomes - tab separated in 2 or 3 columns (FASTA file, genome ID, translation table [optional])</p>
</dd>
</dl>
</div>
<div class="section" id="required named arguments">
<h3>required named arguments<a class="headerlink" href="#required named arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--out_dir</kbd></dt>
<dd><p>directory to output files</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--skip_ani_screen</kbd></dt>
<dd><p>Skip the skani ANI screening step to classify genomes.</p>
</dd>
<dt><kbd>-f, --full_tree</kbd></dt>
<dd><p>use the unsplit bacterial tree for the classify step; this is the original GTDB-Tk approach (version &lt; 2) and requires more than 320 GB of RAM to load the reference tree</p>
</dd>
<dt><kbd>-x, --extension</kbd></dt>
<dd><p>extension of files to process, <code class="docutils literal notranslate"><span class="pre">gz</span></code> = gzipped</p>
<p>Default: “fna”</p>
</dd>
<dt><kbd>--min_perc_aa</kbd></dt>
<dd><p>exclude genomes that do not have at least this percentage of AA in the MSA (inclusive bound)</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--prefix</kbd></dt>
<dd><p>prefix for all output files</p>
<p>Default: “gtdbtk”</p>
</dd>
<dt><kbd>--genes</kbd></dt>
<dd><p>indicates input files contain predicted proteins as amino acids (skip gene calling).Warning: This flag will skip the ANI comparison steps (ANI screen and classification).</p>
</dd>
<dt><kbd>--cpus</kbd></dt>
<dd><p>number of CPUs to use</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--pplacer_cpus</kbd></dt>
<dd><p>number of CPUs to use during pplacer placement</p>
</dd>
<dt><kbd>--force</kbd></dt>
<dd><p>continue processing if an error occurs on a single genome</p>
</dd>
<dt><kbd>--scratch_dir</kbd></dt>
<dd><p>reduce pplacer memory usage by writing to disk (slower).</p>
</dd>
<dt><kbd>--write_single_copy_genes</kbd></dt>
<dd><p>output unaligned single-copy marker genes</p>
</dd>
<dt><kbd>--keep_intermediates</kbd></dt>
<dd><p>keep intermediate files in the final directory</p>
</dd>
<dt><kbd>--min_af</kbd></dt>
<dd><p>minimum alignment fraction to assign genome to a species cluster</p>
<p>Default: 0.5</p>
</dd>
<dt><kbd>--tmpdir</kbd></dt>
<dd><p>specify alternative directory for temporary files</p>
<p>Default: “/tmp”</p>
</dd>
<dt><kbd>--debug</kbd></dt>
<dd><p>create intermediate files for debugging purposes</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gtdbtk<span class="w"> </span>classify_wf<span class="w"> </span>--batchfile<span class="w"> </span>genomes/3_batchfile.tsv<span class="w"> </span>--out_dir<span class="w"> </span>classify_wf_3_genomes<span class="w"> </span>--cpus<span class="w"> </span><span class="m">20</span>
</pre></div>
</div>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>[2025-08-05 20:31:02] INFO: GTDB-Tk v2.5.0
[2025-08-05 20:31:02] INFO: gtdbtk classify_wf --batchfile genomes/3_batchfile.tsv --out_dir classify_wf_3_genomes --cpus 20
[2025-08-05 20:31:02] INFO: Using GTDB-Tk reference data version r226: /srv/db/gtdbtk/official/release226
[2025-08-05 20:31:02] INFO: Loading reference genomes.
[2025-08-05 20:31:03] INFO: Calculating all vs all ANI with skani v0.2.1.
[2025-08-05 20:31:04] INFO: Sketching genomes
[2025-08-05 20:33:38] INFO: Sketches done: 2min 34secs
[2025-08-05 20:33:38] INFO: Running comparisons
[2025-08-05 20:33:57] INFO: Comparisons finished, capturing results.
[2025-08-05 20:35:27] INFO: 0 genome(s) have been classified using the ANI pre-screening step.
[2025-08-05 20:35:27] INFO: Done.
[2025-08-05 20:35:27] INFO: Identifying markers in 3 genomes with 20 threads.
[2025-08-05 20:35:27] TASK: Running Prodigal V2.6.3 to identify genes.
[2025-08-05 20:36:09] INFO: Completed 3 genomes in 41.78 seconds (13.93 seconds/genome).
[2025-08-05 20:36:09] TASK: Identifying TIGRFAM protein families.
[2025-08-05 20:36:14] INFO: Completed 3 genomes in 5.57 seconds (1.86 seconds/genome).
[2025-08-05 20:36:14] TASK: Identifying Pfam protein families.
[2025-08-05 20:36:15] INFO: Completed 3 genomes in 0.70 seconds (4.28 genomes/second).
[2025-08-05 20:36:15] INFO: Annotations done using HMMER 3.1b2 (February 2015).
[2025-08-05 20:36:15] TASK: Summarising identified marker genes.
[2025-08-05 20:36:15] INFO: Completed 3 genomes in 0.17 seconds (17.58 genomes/second).
[2025-08-05 20:36:15] INFO: Done.
[2025-08-05 20:36:15] INFO: Aligning markers in 3 genomes with 20 CPUs.
[2025-08-05 20:36:16] INFO: Processing 3 genomes identified as bacterial.
[2025-08-05 20:36:27] INFO: Read concatenated alignment for 136,646 GTDB genomes.
[2025-08-05 20:36:27] TASK: Generating concatenated alignment for each marker.
[2025-08-05 20:36:33] INFO: Completed 3 genomes in 0.07 seconds (45.19 genomes/second).
[2025-08-05 20:36:34] TASK: Aligning 119 identified markers using hmmalign 3.1b2 (February 2015).
[2025-08-05 20:36:46] INFO: Completed 119 markers in 5.66 seconds (21.02 markers/second).
[2025-08-05 20:36:46] TASK: Masking columns of bacterial multiple sequence alignment using canonical mask.
[2025-08-05 20:40:54] INFO: Completed 136,649 sequences in 4.12 minutes (33,135.45 sequences/minute).
[2025-08-05 20:40:54] INFO: Masked bacterial alignment from 41,084 to 5,036 AAs.
[2025-08-05 20:40:54] INFO: 0 bacterial user genomes have amino acids in &lt;10.0% of columns in filtered MSA.
[2025-08-05 20:40:54] INFO: Creating concatenated alignment for 136,649 bacterial GTDB and user genomes.
[2025-08-05 20:41:36] INFO: Creating concatenated alignment for 3 bacterial user genomes.
[2025-08-05 20:41:38] INFO: Done.
[2025-08-05 20:41:39] TASK: Placing 3 bacterial genomes into backbone reference tree with pplacer using 20 CPUs (be patient).
[2025-08-05 20:41:39] INFO: pplacer version: v1.1.alpha19-0-g807f6f3
[2025-08-05 20:44:58] INFO: Calculating RED values based on reference tree.
[2025-08-05 20:44:59] INFO: 3 out of 3 have an class assignments. Those genomes will be reclassified.
[2025-08-05 20:44:59] TASK: Placing 2 bacterial genomes into class-level reference tree 7 (1/2) with pplacer using 20 CPUs (be patient).
[2025-08-05 20:56:48] INFO: Calculating RED values based on reference tree.
[2025-08-05 20:56:52] TASK: Traversing tree to determine classification method.
[2025-08-05 20:56:52] INFO: Completed 2 genomes in 0.00 seconds (6,026.30 genomes/second).
[2025-08-05 20:56:52] TASK: Calculating average nucleotide identity using skani (v0.2.1).
[2025-08-05 20:56:53] INFO: Completed 13 comparisons in 0.53 seconds (24.35 comparisons/second).
[2025-08-05 20:56:53] INFO: 0 genome(s) have been classified using skani and pplacer.
[2025-08-05 20:56:53] TASK: Placing 1 bacterial genomes into class-level reference tree 2 (2/2) with pplacer using 20 CPUs (be patient).
[2025-08-05 21:08:19] INFO: Calculating RED values based on reference tree.
[2025-08-05 21:08:25] TASK: Traversing tree to determine classification method.
[2025-08-05 21:08:25] INFO: Completed 1 genome in 0.00 seconds (1,531.89 genomes/second).
[2025-08-05 21:08:25] INFO: 0 genome(s) have been classified using skani and pplacer.
[2025-08-05 21:08:25] INFO: Note that Tk classification mode is insufficient for publication of new taxonomic designations. New designations should be based on one or more de novo trees, an example of which can be produced by Tk in de novo mode.
[2025-08-05 21:08:25] INFO: Done.
[2025-08-05 21:08:25] INFO: Removing intermediate files.
[2025-08-05 21:08:25] INFO: Intermediate files removed.
[2025-08-05 21:08:25] INFO: Done.
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="convert_to_itol.html" class="btn btn-neutral float-right" title="convert_to_itol" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="classify.html" class="btn btn-neutral float-left" title="classify" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2025, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>