

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>de_novo_wf &mdash; GTDB-Tk 2.4.1 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/commands/de_novo_wf.html" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="decorate" href="decorate.html" />
    <link rel="prev" title="convert_to_itol" href="convert_to_itol.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Commands</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="align.html">align</a></li>
<li class="toctree-l2"><a class="reference internal" href="ani_rep.html">ani_rep</a></li>
<li class="toctree-l2"><a class="reference internal" href="check_install.html">check_install</a></li>
<li class="toctree-l2"><a class="reference internal" href="classify.html">classify</a></li>
<li class="toctree-l2"><a class="reference internal" href="classify_wf.html">classify_wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="convert_to_itol.html">convert_to_itol</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">de_novo_wf</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#mutually exclusive required arguments">mutually exclusive required arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mutually exclusive required arguments_repeat1">mutually exclusive required arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#required named arguments">required named arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#input">Input</a></li>
<li class="toctree-l4"><a class="reference internal" href="#custom-taxonomy-format">Custom Taxonomy Format</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="decorate.html">decorate</a></li>
<li class="toctree-l2"><a class="reference internal" href="export_msa.html">export_msa</a></li>
<li class="toctree-l2"><a class="reference internal" href="identify.html">identify</a></li>
<li class="toctree-l2"><a class="reference internal" href="infer.html">infer</a></li>
<li class="toctree-l2"><a class="reference internal" href="root.html">root</a></li>
<li class="toctree-l2"><a class="reference internal" href="test.html">test</a></li>
<li class="toctree-l2"><a class="reference internal" href="trim_msa.html">trim_msa</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">Commands</a> &raquo;</li>
        
      <li>de_novo_wf</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/commands/de_novo_wf.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="de-novo-wf">
<span id="commands-de-novo-wf"></span><h1>de_novo_wf<a class="headerlink" href="#de-novo-wf" title="Permalink to this headline">Â¶</a></h1>
<p>For arguments and output files, see each of the individual steps:</p>
<ul class="simple">
<li><p><a class="reference internal" href="identify.html#commands-identify"><span class="std std-ref">identify</span></a></p></li>
<li><p><a class="reference internal" href="align.html#commands-align"><span class="std std-ref">align</span></a></p></li>
<li><p><a class="reference internal" href="infer.html#commands-infer"><span class="std std-ref">infer</span></a></p></li>
<li><p><a class="reference internal" href="root.html#commands-root"><span class="std std-ref">root</span></a></p></li>
<li><p><a class="reference internal" href="decorate.html#commands-decorate"><span class="std std-ref">decorate</span></a></p></li>
</ul>
<p>The <em>de novo</em> workflow infers new bacterial and archaeal trees containing all user supplied and GTDB-Tk reference genomes.
The classify workflow is recommended for obtaining taxonomic classifications, and this workflow only recommended if
a <em>de novo</em> domain-specific trees are desired. <strong>One should take the taxonomic assignments as a guide, but not as final classifications</strong>. In particular, no effort is made to resolve the taxonomic assignment of lineages composed exclusively of user submitted genomes.</p>
<p>This workflow consists of five steps: <code class="docutils literal notranslate"><span class="pre">identify</span></code>, <code class="docutils literal notranslate"><span class="pre">align</span></code>, <code class="docutils literal notranslate"><span class="pre">infer</span></code>, <code class="docutils literal notranslate"><span class="pre">root</span></code>,
and <code class="docutils literal notranslate"><span class="pre">decorate</span></code>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">identify</span></code> and <code class="docutils literal notranslate"><span class="pre">align</span></code> steps are the same as in the classify workflow.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">infer</span></code> step uses <a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a> with the WAG+GAMMA models to calculate independent, <em>de novo</em> bacterial and archaeal trees.
These trees can then be rooted using a user specified outgroup and decorated with the GTDB taxonomy.</p>
<p>The <em>de novo</em> workflow can be run as follows:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gtdbtk<span class="w"> </span>de_novo_wf<span class="w"> </span>--genome_dir<span class="w"> </span>&lt;my_genomes&gt;<span class="w"> </span>--&lt;marker_set&gt;<span class="w"> </span>--outgroup_taxon<span class="w"> </span>&lt;outgroup&gt;<span class="w"> </span>--out_dir<span class="w"> </span>&lt;output_dir&gt;
</pre></div>
</div>
<p>This will process all genomes in &lt;my_genomes&gt; using the specified marker set and place the results in &lt;output_dir&gt;.
Only genomes previously identified as being bacterial (archaeal) should be included when using the bacterial (archaeal) marker set.
The tree will be rooted with the &lt;outgroup&gt; taxon (typically a phylum in the domain-specific tree) as required for
correct decoration of the tree. In general, we suggest the resulting tree be treated as unrooted when interpreting results.
Identical to the classify workflow, the location of genomes can also be specified using a batch file with the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> flag.</p>
<p>The workflow supports several optional flags, including:</p>
<ul class="simple">
<li><p>cpus: maximum number of CPUs to use</p></li>
<li><p>min_perc_aa: filter genomes with an insufficient percentage of AA in the MSA (default: 50)</p></li>
<li><p>taxa_filter: filter genomes to taxa within specific taxonomic groups</p></li>
<li><p>prot_model: protein substitution model for tree inference (LG or WAG; default: WAG)</p></li>
</ul>
<p>For other flags please consult the command line interface.</p>
<div class="section" id="arguments">
<h2>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">Â¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">gtdbtk</span> <span class="n">de_novo_wf</span> <span class="p">(</span><span class="o">--</span><span class="n">genome_dir</span> <span class="n">GENOME_DIR</span> <span class="o">|</span> <span class="o">--</span><span class="n">batchfile</span> <span class="n">BATCHFILE</span><span class="p">)</span>
                         <span class="p">(</span><span class="o">--</span><span class="n">bacteria</span> <span class="o">|</span> <span class="o">--</span><span class="n">archaea</span><span class="p">)</span> <span class="o">--</span><span class="n">outgroup_taxon</span>
                         <span class="n">OUTGROUP_TAXON</span> <span class="o">--</span><span class="n">out_dir</span> <span class="n">OUT_DIR</span> <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">EXTENSION</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">skip_gtdb_refs</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">taxa_filter</span> <span class="n">TAXA_FILTER</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">min_perc_aa</span> <span class="n">MIN_PERC_AA</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">custom_msa_filters</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">cols_per_gene</span> <span class="n">COLS_PER_GENE</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">min_consensus</span> <span class="n">MIN_CONSENSUS</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">max_consensus</span> <span class="n">MAX_CONSENSUS</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">min_perc_taxa</span> <span class="n">MIN_PERC_TAXA</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">rnd_seed</span> <span class="n">RND_SEED</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">prot_model</span> <span class="p">{</span><span class="n">JTT</span><span class="p">,</span><span class="n">WAG</span><span class="p">,</span><span class="n">LG</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_support</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gamma</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">gtdbtk_classification_file</span> <span class="n">GTDBTK_CLASSIFICATION_FILE</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">custom_taxonomy_file</span> <span class="n">CUSTOM_TAXONOMY_FILE</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">write_single_copy_genes</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">prefix</span> <span class="n">PREFIX</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">genes</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">cpus</span> <span class="n">CPUS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">force</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tmpdir</span> <span class="n">TMPDIR</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">--</span><span class="n">keep_intermediates</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">debug</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="mutually exclusive required arguments">
<h3>mutually exclusive required arguments<a class="headerlink" href="#mutually exclusive required arguments" title="Permalink to this headline">Â¶</a></h3>
<dl class="option-list">
<dt><kbd>--genome_dir</kbd></dt>
<dd><p>directory containing genome files in FASTA format</p>
</dd>
<dt><kbd>--batchfile</kbd></dt>
<dd><p>path to file describing genomes - tab separated in 2 or 3 columns (FASTA file, genome ID, translation table [optional])</p>
</dd>
</dl>
</div>
<div class="section" id="mutually exclusive required arguments_repeat1">
<h3>mutually exclusive required arguments<a class="headerlink" href="#mutually exclusive required arguments_repeat1" title="Permalink to this headline">Â¶</a></h3>
<dl class="option-list">
<dt><kbd>--bacteria</kbd></dt>
<dd><p>process bacterial genomes</p>
</dd>
<dt><kbd>--archaea</kbd></dt>
<dd><p>process archaeal genomes</p>
</dd>
</dl>
</div>
<div class="section" id="required named arguments">
<h3>required named arguments<a class="headerlink" href="#required named arguments" title="Permalink to this headline">Â¶</a></h3>
<dl class="option-list">
<dt><kbd>--outgroup_taxon</kbd></dt>
<dd><p>taxon to use as outgroup (e.g., <code class="docutils literal notranslate"><span class="pre">p__Patescibacteria</span></code> or <code class="docutils literal notranslate"><span class="pre">p__Altiarchaeota</span></code>)</p>
</dd>
<dt><kbd>--out_dir</kbd></dt>
<dd><p>directory to output files</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">Â¶</a></h3>
<dl class="option-list">
<dt><kbd>-x, --extension</kbd></dt>
<dd><p>extension of files to process, <code class="docutils literal notranslate"><span class="pre">gz</span></code> = gzipped</p>
<p>Default: âfnaâ</p>
</dd>
<dt><kbd>--skip_gtdb_refs</kbd></dt>
<dd><p>do not include GTDB reference genomes in multiple sequence alignment.</p>
</dd>
<dt><kbd>--taxa_filter</kbd></dt>
<dd><p>filter GTDB genomes to taxa (comma separated) within specific taxonomic groups (e.g.: <code class="docutils literal notranslate"><span class="pre">d__Bacteria</span></code> or <code class="docutils literal notranslate"><span class="pre">p__Proteobacteria,p__Actinobacteria</span></code>)</p>
</dd>
<dt><kbd>--min_perc_aa</kbd></dt>
<dd><p>exclude genomes that do not have at least this percentage of AA in the MSA (inclusive bound)</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--custom_msa_filters</kbd></dt>
<dd><p>perform custom filtering of MSA with <code class="docutils literal notranslate"><span class="pre">cols_per_gene</span></code>, <code class="docutils literal notranslate"><span class="pre">min_consensus</span></code> <code class="docutils literal notranslate"><span class="pre">max_consensus</span></code>, and <code class="docutils literal notranslate"><span class="pre">min_perc_taxa</span></code> parameters instead of using canonical mask</p>
</dd>
<dt><kbd>--cols_per_gene</kbd></dt>
<dd><p>maximum number of columns to retain per gene when generating the MSA</p>
<p>Default: 42</p>
</dd>
<dt><kbd>--min_consensus</kbd></dt>
<dd><p>minimum percentage of the same amino acid required to retain column (inclusive bound)</p>
<p>Default: 25</p>
</dd>
<dt><kbd>--max_consensus</kbd></dt>
<dd><p>maximum percentage of the same amino acid required to retain column (exclusive bound)</p>
<p>Default: 95</p>
</dd>
<dt><kbd>--min_perc_taxa</kbd></dt>
<dd><p>minimum percentage of taxa required to retain column (inclusive bound)</p>
<p>Default: 50</p>
</dd>
<dt><kbd>--rnd_seed</kbd></dt>
<dd><p>random seed to use for selecting columns, e.g. <code class="docutils literal notranslate"><span class="pre">42</span></code></p>
</dd>
<dt><kbd>--prot_model</kbd></dt>
<dd><p>Possible choices: JTT, WAG, LG</p>
<p>protein substitution model for tree inference</p>
<p>Default: âWAGâ</p>
</dd>
<dt><kbd>--no_support</kbd></dt>
<dd><p>do not compute local support values using the Shimodaira-Hasegawa test</p>
</dd>
<dt><kbd>--gamma</kbd></dt>
<dd><p>rescale branch lengths to optimize the Gamma20 likelihood</p>
</dd>
<dt><kbd>--gtdbtk_classification_file</kbd></dt>
<dd><p>file with GTDB-Tk classifications produced by the <cite>classify</cite> command</p>
</dd>
<dt><kbd>--custom_taxonomy_file</kbd></dt>
<dd><p>file indicating custom taxonomy strings for user genomes, that should contain any genomes belonging to the outgroup. Format: GENOME_ID&lt;TAB&gt;d__;p__;c__;o__;f__;g__;s__</p>
</dd>
<dt><kbd>--write_single_copy_genes</kbd></dt>
<dd><p>output unaligned single-copy marker genes</p>
</dd>
<dt><kbd>--prefix</kbd></dt>
<dd><p>prefix for all output files</p>
<p>Default: âgtdbtkâ</p>
</dd>
<dt><kbd>--genes</kbd></dt>
<dd><p>indicates input files contain predicted proteins as amino acids (skip gene calling).Warning: This flag will skip the ANI comparison steps (ani_screen and classification).</p>
</dd>
<dt><kbd>--cpus</kbd></dt>
<dd><p>number of CPUs to use</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--force</kbd></dt>
<dd><p>continue processing if an error occurs on a single genome</p>
</dd>
<dt><kbd>--tmpdir</kbd></dt>
<dd><p>specify alternative directory for temporary files</p>
<p>Default: â/tmpâ</p>
</dd>
<dt><kbd>--keep_intermediates</kbd></dt>
<dd><p>keep intermediate files in the final directory</p>
</dd>
<dt><kbd>--debug</kbd></dt>
<dd><p>create intermediate files for debugging purposes</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">Â¶</a></h2>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gtdbtk<span class="w"> </span>de_novo_wf<span class="w"> </span>--genome_dir<span class="w"> </span>genomes/<span class="w"> </span>--outgroup_taxon<span class="w"> </span>p__Undinarchaeota<span class="w"> </span>--archaea<span class="w"> </span>--out_dir<span class="w"> </span>de_novo_wf<span class="w"> </span>--cpus<span class="w"> </span><span class="m">3</span>

gtdbtk<span class="w"> </span>de_novo_wf<span class="w"> </span>--genome_dir<span class="w"> </span>genomes/<span class="w"> </span>--outgroup_taxon<span class="w"> </span>p__Chloroflexota<span class="w"> </span>--bacteria<span class="w">  </span>--taxa_filter<span class="w"> </span>p__Firmicutes<span class="w"> </span>--out_dir<span class="w"> </span>de_novo_output

<span class="c1">#Skip GTDB reference genomes ( requires --custom_taxonomy_file for outgrouping)</span>
gtdbtk<span class="w"> </span>de_novo_wf<span class="w"> </span>--genome_dir<span class="w"> </span>genomes/<span class="w"> </span>--outgroup_taxon<span class="w"> </span>p__Customphylum<span class="w"> </span>--bacteria<span class="w"> </span>--custom_taxonomy_file<span class="w"> </span>custom_taxonomy.tsv<span class="w"> </span>--out_dir<span class="w"> </span>de_novo_output

<span class="c1">#Use a subset of GTDB reference genomes (p__Firmicutes) and outgroup on a custom Phylum (p__Customphylum)</span>
gtdbtk<span class="w"> </span>de_novo_wf<span class="w"> </span>--genome_dir<span class="w"> </span>genomes/<span class="w"> </span>--taxa_filter<span class="w"> </span>p__Firmicutes<span class="w"> </span>--outgroup_taxon<span class="w"> </span>p__Customphylum<span class="w"> </span>--bacteria<span class="w"> </span>--custom_taxonomy_file<span class="w"> </span>custom_taxonomy.tsv<span class="w"> </span>--out_dir<span class="w"> </span>de_novo_output
</pre></div>
</div>
</div>
<div class="section" id="custom-taxonomy-format">
<h3>Custom Taxonomy Format<a class="headerlink" href="#custom-taxonomy-format" title="Permalink to this headline">Â¶</a></h3>
<p>The custom taxonomy file is a Tab-delimited file with the first column listing user genomes (i.e Fasta filename without the extension)
and the second column listing the standardized 7-rank taxonomy.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">#For genome_1.fna, genome_2.fna and genome_3.fna</span>
genome_1<span class="w">    </span>d__Bacteria<span class="p">;</span>p__Proteobacteria<span class="p">;</span>c__Gammaproteobacteria<span class="p">;</span>o__Enterobacterales<span class="p">;</span>f__Enterobacteriaceae<span class="p">;</span>g__Salmonella<span class="p">;</span>s__Salmonella<span class="w"> </span>enterica
genome_2<span class="w">    </span>d__Bacteria<span class="p">;</span>p__Actinobacteriota<span class="p">;</span>c__Actinomycetia<span class="p">;</span>o__Mycobacteriales<span class="p">;</span>f__Mycobacteriaceae<span class="p">;</span>g__Mycobacterium<span class="p">;</span>s__Mycobacterium<span class="w"> </span>tuberculosis
genome_3<span class="w">    </span>d__Bacteria<span class="p">;</span>p__Firmicutes<span class="p">;</span>c__Bacilli<span class="p">;</span>o__Lactobacillales<span class="p">;</span>f__Streptococcaceae<span class="p">;</span>g__Streptococcus<span class="p">;</span>s__Streptococcus<span class="w"> </span>pyogenes
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="decorate.html" class="btn btn-neutral float-right" title="decorate" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="convert_to_itol.html" class="btn btn-neutral float-left" title="convert_to_itol" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2025, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>