

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>align &mdash; GTDB-Tk 1.4.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script src="../_static/custom.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="ani_rep" href="ani_rep.html" />
    <link rel="prev" title="Commands" href="index.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.4.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption"><span class="caption-text">Running GTDB-Tk</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Commands</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">align</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#required named arguments">required named arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mutually exclusive optional arguments">mutually exclusive optional arguments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#files-output">Files output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#input">Input</a></li>
<li class="toctree-l4"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="ani_rep.html">ani_rep</a></li>
<li class="toctree-l2"><a class="reference internal" href="check_install.html">check_install</a></li>
<li class="toctree-l2"><a class="reference internal" href="classify.html">classify</a></li>
<li class="toctree-l2"><a class="reference internal" href="classify_wf.html">classify_wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="de_novo_wf.html">de_novo_wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="decorate.html">decorate</a></li>
<li class="toctree-l2"><a class="reference internal" href="export_msa.html">export_msa</a></li>
<li class="toctree-l2"><a class="reference internal" href="identify.html">identify</a></li>
<li class="toctree-l2"><a class="reference internal" href="infer.html">infer</a></li>
<li class="toctree-l2"><a class="reference internal" href="root.html">root</a></li>
<li class="toctree-l2"><a class="reference internal" href="test.html">test</a></li>
<li class="toctree-l2"><a class="reference internal" href="trim_msa.html">trim_msa</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">Commands</a> &raquo;</li>
        
      <li>align</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs_src/src/commands/align.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="align">
<span id="commands-align"></span><h1>align<a class="headerlink" href="#align" title="Permalink to this headline">¶</a></h1>
<p>Create a multiple sequence alignment based on the AR122/BAC120 marker set.</p>
<div class="section" id="arguments">
<h2>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">gtdbtk</span> <span class="n">align</span> <span class="o">--</span><span class="n">identify_dir</span> <span class="n">IDENTIFY_DIR</span> <span class="o">--</span><span class="n">out_dir</span> <span class="n">OUT_DIR</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">skip_gtdb_refs</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">taxa_filter</span> <span class="n">TAXA_FILTER</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">min_perc_aa</span> <span class="n">MIN_PERC_AA</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">cols_per_gene</span> <span class="n">COLS_PER_GENE</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">min_consensus</span> <span class="n">MIN_CONSENSUS</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">max_consensus</span> <span class="n">MAX_CONSENSUS</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">min_perc_taxa</span> <span class="n">MIN_PERC_TAXA</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">rnd_seed</span> <span class="n">RND_SEED</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">prefix</span> <span class="n">PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">cpus</span> <span class="n">CPUS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">debug</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">custom_msa_filters</span> <span class="o">|</span> <span class="o">--</span><span class="n">skip_trimming</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="required named arguments">
<h3>required named arguments<a class="headerlink" href="#required named arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--identify_dir</kbd></dt>
<dd><p>output directory of ‘identify’ command</p>
</dd>
<dt><kbd>--out_dir</kbd></dt>
<dd><p>directory to output files</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--skip_gtdb_refs</kbd></dt>
<dd><p>do not include GTDB reference genomes in multiple sequence alignment</p>
</dd>
<dt><kbd>--taxa_filter</kbd></dt>
<dd><p>filter GTDB genomes to taxa (comma separated) within specific taxonomic groups (e.g.: <code class="docutils literal notranslate"><span class="pre">d__Bacteria</span></code> or <code class="docutils literal notranslate"><span class="pre">p__Proteobacteria,p__Actinobacteria</span></code>)</p>
</dd>
<dt><kbd>--min_perc_aa</kbd></dt>
<dd><p>exclude genomes that do not have at least this percentage of AA in the MSA (inclusive bound)</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--cols_per_gene</kbd></dt>
<dd><p>maximum number of columns to retain per gene when generating the MSA</p>
<p>Default: 42</p>
</dd>
<dt><kbd>--min_consensus</kbd></dt>
<dd><p>minimum percentage of the same amino acid required to retain column (inclusive bound)</p>
<p>Default: 25</p>
</dd>
<dt><kbd>--max_consensus</kbd></dt>
<dd><p>maximum percentage of the same amino acid required to retain column (exclusive bound)</p>
<p>Default: 95</p>
</dd>
<dt><kbd>--min_perc_taxa</kbd></dt>
<dd><p>minimum percentage of taxa required to retain column (inclusive bound)</p>
<p>Default: 50</p>
</dd>
<dt><kbd>--rnd_seed</kbd></dt>
<dd><p>random seed to use for selecting columns, e.g. <code class="docutils literal notranslate"><span class="pre">42</span></code></p>
</dd>
<dt><kbd>--prefix</kbd></dt>
<dd><p>prefix for all output files</p>
<p>Default: “gtdbtk”</p>
</dd>
<dt><kbd>--cpus</kbd></dt>
<dd><p>number of CPUs to use</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--debug</kbd></dt>
<dd><p>create intermediate files for debugging purposes</p>
</dd>
</dl>
</div>
<div class="section" id="mutually exclusive optional arguments">
<h3>mutually exclusive optional arguments<a class="headerlink" href="#mutually exclusive optional arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--custom_msa_filters</kbd></dt>
<dd><p>perform custom filtering of MSA with <code class="docutils literal notranslate"><span class="pre">cols_per_gene</span></code>, <code class="docutils literal notranslate"><span class="pre">min_consensus</span></code> <code class="docutils literal notranslate"><span class="pre">max_consensus</span></code>, and <code class="docutils literal notranslate"><span class="pre">min_perc_taxa</span></code> parameters instead of using canonical mask</p>
</dd>
<dt><kbd>--skip_trimming</kbd></dt>
<dd><p>skip the trimming step and return the full MSAs</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="files-output">
<h2>Files output<a class="headerlink" href="#files-output" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="../files/filtered.tsv.html#files-filtered-tsv"><span class="std std-ref">[prefix].[domain].filtered.tsv</span></a></p></li>
<li><p><a class="reference internal" href="../files/msa.fasta.html#files-msa-fasta"><span class="std std-ref">[prefix].[domain].msa.fasta</span></a></p></li>
<li><p><a class="reference internal" href="../files/user_msa.fasta.html#files-user-msa-fasta"><span class="std std-ref">[prefix].[domain].user_msa.fasta</span></a></p></li>
<li><p><a class="reference internal" href="../files/gtdbtk.log.html#files-gtdbtk-log"><span class="std std-ref">[prefix].log</span></a></p></li>
<li><p><a class="reference internal" href="../files/gtdbtk.warnings.log.html#files-gtdbtk-warnings-log"><span class="std std-ref">[prefix].warnings.log</span></a></p></li>
<li><p><a class="reference internal" href="../files/marker_info.tsv.html#files-marker-info-tsv"><span class="std std-ref">align/intermediate_results/[prefix].[domain].marker_info.tsv</span></a></p></li>
</ul>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gtdbtk align --identify_dir identify_output/ --out_dir align_output --cpus <span class="m">3</span>
</pre></div>
</div>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>[2020-04-14 09:14:44] INFO: GTDB-Tk v1.1.0
[2020-04-14 09:14:44] INFO: gtdbtk align --identify_dir identify_output/ --out_dir align_output --cpus 3
[2020-04-14 09:14:44] INFO: Using GTDB-Tk reference data version r89: /release89
[2020-04-14 09:14:44] INFO: Aligning markers in 3 genomes with 3 threads.
[2020-04-14 09:14:44] INFO: Processing 3 genomes identified as archaeal.
[2020-04-14 09:14:44] INFO: Read concatenated alignment for 1248 GTDB genomes.
==&gt; Finished aligning 3 of 3 (100.0%) genomes.
[2020-04-14 09:14:49] INFO: Masking columns of multiple sequence alignment using canonical mask.
[2020-04-14 09:14:52] INFO: Masked alignment from 32675 to 5124 AAs.
[2020-04-14 09:14:52] INFO: 0 user genomes have amino acids in &lt;10.0% of columns in filtered MSA.
[2020-04-14 09:14:52] INFO: Creating concatenated alignment for 1251 GTDB and user genomes.
[2020-04-14 09:14:52] INFO: Creating concatenated alignment for 3 user genomes.
[2020-04-14 09:14:52] INFO: Done.
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="ani_rep.html" class="btn btn-neutral float-right" title="ani_rep" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="Commands" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Pierre-Alain Chaumeil, Aaron Mussig, and Donovan Parks

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>