

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>gtdbtk.json &mdash; GTDB-Tk 2.5.1 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/files/gtdbtk.json.html" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="failed.genomes.tsv" href="failed_genomes.tsv.html" />
    <link rel="prev" title="tree.mapping.tsv" href="tree.mapping.tsv.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.5.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands/index.html">Commands</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Files</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="ani_closest.tsv.html">ani_closest.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="ani_summary.tsv.html">ani_summary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="classification_pplacer.tsv.html">classification_pplacer.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="classify.tree.html">classify.tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="fasttree.log.html">fasttree.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="filtered.tsv.html">filtered.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="gtdbtk.log.html">gtdbtk.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="gtdbtk.warnings.log.html">gtdbtk.warnings.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="gtdbtk_ref_sketch.msh.html">gtdbtk_ref_sketch.msh</a></li>
<li class="toctree-l2"><a class="reference internal" href="marker_info.tsv.html">marker_info.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="markers_summary.tsv.html">markers_summary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="msa.fasta.html">msa.fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="pfam.tsv.html">pfam.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="pfam_tophit.tsv.html">pfam_tophit.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="pplacer.domain.json.html">pplacer.&lt;domain&gt;.json</a></li>
<li class="toctree-l2"><a class="reference internal" href="pplacer.domain.out.html">pplacer.&lt;domain&gt;.out</a></li>
<li class="toctree-l2"><a class="reference internal" href="prodigal_translation_table.tsv.html">prodigal_translation_table.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="protein.faa.html">protein.faa</a></li>
<li class="toctree-l2"><a class="reference internal" href="protein.fna.html">protein.fna</a></li>
<li class="toctree-l2"><a class="reference internal" href="protein.gff.html">protein.gff</a></li>
<li class="toctree-l2"><a class="reference internal" href="red_dictionary.tsv.html">red_dictionary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="summary.tsv.html">summary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="test_execution.log.html">test_execution.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="tigrfam.out.html">tigrfam.out</a></li>
<li class="toctree-l2"><a class="reference internal" href="tigrfam.tsv.html">tigrfam.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="tigrfam_tophit.tsv.html">tigrfam_tophit.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="translation_table_summary.tsv.html">translation_table_summary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="tree.log.html">tree.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="unrooted.tree.html">unrooted.tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="user_msa.fasta.html">user_msa.fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="user_query_sketch.msh.html">user_query_sketch.msh</a></li>
<li class="toctree-l2"><a class="reference internal" href="tree.mapping.tsv.html">tree.mapping.tsv</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">gtdbtk.json</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#produced-by">Produced by</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="failed_genomes.tsv.html">failed.genomes.tsv</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">Files</a> &raquo;</li>
        
      <li>gtdbtk.json</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/files/gtdbtk.json.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="gtdbtk-json">
<span id="files-gtdbtk-json"></span><h1>gtdbtk.json<a class="headerlink" href="#gtdbtk-json" title="Permalink to this headline">¶</a></h1>
<p>The console output of GTDB-Tk saved to disk in a JSON format.</p>
<div class="section" id="produced-by">
<h2>Produced by<a class="headerlink" href="#produced-by" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="../commands/align.html#commands-align"><span class="std std-ref">align</span></a></p></li>
<li><p><a class="reference internal" href="../commands/align.html#commands-align"><span class="std std-ref">align</span></a></p></li>
<li><p><a class="reference internal" href="../commands/classify.html#commands-classify"><span class="std std-ref">classify</span></a></p></li>
<li><p><a class="reference internal" href="../commands/classify_wf.html#commands-classify-wf"><span class="std std-ref">classify_wf</span></a></p></li>
<li><p><a class="reference internal" href="../commands/de_novo_wf.html#commands-de-novo-wf"><span class="std std-ref">de_novo_wf</span></a></p></li>
<li><p><a class="reference internal" href="../commands/identify.html#commands-identify"><span class="std std-ref">identify</span></a></p></li>
<li><p><a class="reference internal" href="../commands/infer.html#commands-infer"><span class="std std-ref">infer</span></a></p></li>
</ul>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>{
&quot;version&quot;: &quot;2.5.0&quot;,
&quot;command_line&quot;: &quot;gtdbtk classify_wf --batchfile genomes/3_batchfile.tsv --out_dir classify_wf_3_genomes --cpus 20&quot;,
&quot;database_version&quot;: &quot;r226&quot;,
&quot;database_path&quot;: &quot;/srv/db/gtdbtk/official/release226&quot;,
&quot;steps&quot;: [
    {
        &quot;name&quot;: &quot;ANI screen&quot;,
        &quot;output_dir&quot;: &quot;classify_wf_3_genomes&quot;,
        &quot;starts_at&quot;: &quot;2025-08-05T20:31:02.156781&quot;,
        &quot;ends_at&quot;: &quot;2025-08-05T20:35:27.198873&quot;,
        &quot;duration&quot;: &quot;0:04:25&quot;,
        &quot;status&quot;: &quot;completed&quot;,
        &quot;genome_dir&quot;: null,
        &quot;batchfile&quot;: &quot;genomes/3_batchfile.tsv&quot;,
        &quot;min_af&quot;: 0.5,
        &quot;output_files&quot;: {}
    },
    {
        &quot;name&quot;: &quot;identify&quot;,
        &quot;output_dir&quot;: &quot;classify_wf_3_genomes&quot;,
        &quot;starts_at&quot;: &quot;2025-08-05T20:35:27.270095&quot;,
        &quot;ends_at&quot;: &quot;2025-08-05T20:36:15.865020&quot;,
        &quot;duration&quot;: &quot;0:00:48&quot;,
        &quot;status&quot;: &quot;completed&quot;,
        &quot;genes&quot;: false,
        &quot;extension&quot;: &quot;fna&quot;,
        &quot;write_single_copy_genes&quot;: false,
        &quot;genome_dir&quot;: null,
        &quot;batchfile&quot;: &quot;genomes/3_batchfile.tsv&quot;,
        &quot;output_files&quot;: {
            &quot;all&quot;: [
                &quot;classify_wf_3_genomes/identify/gtdbtk.failed_genomes.tsv&quot;,
                &quot;classify_wf_3_genomes/identify/gtdbtk.translation_table_summary.tsv&quot;
            ],
            &quot;ar53&quot;: [
                &quot;classify_wf_3_genomes/identify/gtdbtk.ar53.markers_summary.tsv&quot;
            ],
            &quot;bac120&quot;: [
                &quot;classify_wf_3_genomes/identify/gtdbtk.bac120.markers_summary.tsv&quot;
            ]
        }
    },
    {
        &quot;name&quot;: &quot;align&quot;,
        &quot;output_dir&quot;: &quot;classfiy_wf_3_genomes&quot;,
        &quot;starts_at&quot;: &quot;2025-08-05T20:36:15.865242&quot;,
        &quot;ends_at&quot;: &quot;2025-08-05T20:41:38.376388&quot;,
        &quot;duration&quot;: &quot;0:05:22&quot;,
        &quot;status&quot;: &quot;completed&quot;,
        &quot;identify_dir&quot;: &quot;classfiy_wf_3_genomes&quot;,
        &quot;skip_gtdb_refs&quot;: false,
        &quot;taxa_filter&quot;: null,
        &quot;min_perc_aa&quot;: 10,
        &quot;custom_msa_filters&quot;: false,
        &quot;skip_trimming&quot;: false,
        &quot;rnd_seed&quot;: null,
        &quot;cols_per_gene&quot;: null,
        &quot;min_consensus&quot;: null,
        &quot;max_consensus&quot;: null,
        &quot;min_perc_taxa&quot;: null,
        &quot;outgroup_taxon&quot;: null,
        &quot;output_files&quot;: {
            &quot;bac120&quot;: [
                &quot;classfiy_wf_3_genomes/align/gtdbtk.bac120.filtered.tsv&quot;,
                &quot;classfiy_wf_3_genomes/align/gtdbtk.bac120.msa.fasta&quot;,
                &quot;classfiy_wf_3_genomes/align/gtdbtk.bac120.user_msa.fasta&quot;
            ]
        }
    },
    {
        &quot;name&quot;: &quot;classify&quot;,
        &quot;output_dir&quot;: &quot;classfiy_wf_3_genomes&quot;,
        &quot;starts_at&quot;: &quot;2025-08-05T20:41:38.377084&quot;,
        &quot;ends_at&quot;: &quot;2025-08-05T21:08:25.704484&quot;,
        &quot;duration&quot;: &quot;0:26:47&quot;,
        &quot;status&quot;: &quot;completed&quot;,
        &quot;align_dir&quot;: &quot;classfiy_wf_3_genomes&quot;,
        &quot;genome_dir&quot;: null,
        &quot;batchfile&quot;: &quot;genomes/3_batchfile.tsv&quot;,
        &quot;scratch_dir&quot;: null,
        &quot;debug_option&quot;: false,
        &quot;full_tree&quot;: false,
        &quot;skip_ani_screen&quot;: true,
        &quot;output_files&quot;: {
            &quot;bac120&quot;: [
                &quot;classify_wf_3_genomes/classify/gtdbtk.backbone.bac120.classify.tree&quot;,
                &quot;classify_wf_3_genomes/classify/gtdbtk.bac120.classify.tree.7.tree&quot;,
                &quot;classify_wf_3_genomes/classify/gtdbtk.bac120.classify.tree.2.tree&quot;,
                &quot;classify_wf_3_genomes/classify/gtdbtk.bac120.tree.mapping.tsv&quot;,
                &quot;classify_wf_3_genomes/classify/gtdbtk.bac120.summary.tsv&quot;
            ]
        }
    }
],
&quot;output_dir&quot;: &quot;classify_wf_3_genomes&quot;,
&quot;path&quot;: &quot;classify_wf_3_genomes/gtdbtk.json&quot;
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="failed_genomes.tsv.html" class="btn btn-neutral float-right" title="failed.genomes.tsv" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="tree.mapping.tsv.html" class="btn btn-neutral float-left" title="tree.mapping.tsv" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2025, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>