

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>summary.tsv &mdash; GTDB-Tk 2.4.1 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/files/summary.tsv.html" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="test_execution.log" href="test_execution.log.html" />
    <link rel="prev" title="red_dictionary.tsv" href="red_dictionary.tsv.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.4.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands/index.html">Commands</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Files</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="ani_closest.tsv.html">ani_closest.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="ani_summary.tsv.html">ani_summary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="classification_pplacer.tsv.html">classification_pplacer.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="classify.tree.html">classify.tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="fasttree.log.html">fasttree.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="filtered.tsv.html">filtered.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="gtdbtk.log.html">gtdbtk.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="gtdbtk.warnings.log.html">gtdbtk.warnings.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="gtdbtk_ref_sketch.msh.html">gtdbtk_ref_sketch.msh</a></li>
<li class="toctree-l2"><a class="reference internal" href="marker_info.tsv.html">marker_info.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="markers_summary.tsv.html">markers_summary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="mash_distances.msh.html">mash_distances.msh</a></li>
<li class="toctree-l2"><a class="reference internal" href="msa.fasta.html">msa.fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="pfam.tsv.html">pfam.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="pfam_tophit.tsv.html">pfam_tophit.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="pplacer.domain.json.html">pplacer.&lt;domain&gt;.json</a></li>
<li class="toctree-l2"><a class="reference internal" href="pplacer.domain.out.html">pplacer.&lt;domain&gt;.out</a></li>
<li class="toctree-l2"><a class="reference internal" href="prodigal_translation_table.tsv.html">prodigal_translation_table.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="protein.faa.html">protein.faa</a></li>
<li class="toctree-l2"><a class="reference internal" href="protein.fna.html">protein.fna</a></li>
<li class="toctree-l2"><a class="reference internal" href="protein.gff.html">protein.gff</a></li>
<li class="toctree-l2"><a class="reference internal" href="red_dictionary.tsv.html">red_dictionary.tsv</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">summary.tsv</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#produced-by">Produced by</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="test_execution.log.html">test_execution.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="tigrfam.out.html">tigrfam.out</a></li>
<li class="toctree-l2"><a class="reference internal" href="tigrfam.tsv.html">tigrfam.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="tigrfam_tophit.tsv.html">tigrfam_tophit.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="translation_table_summary.tsv.html">translation_table_summary.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="tree.log.html">tree.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="unrooted.tree.html">unrooted.tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="user_msa.fasta.html">user_msa.fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="user_query_sketch.msh.html">user_query_sketch.msh</a></li>
<li class="toctree-l2"><a class="reference internal" href="tree.mapping.tsv.html">tree.mapping.tsv</a></li>
<li class="toctree-l2"><a class="reference internal" href="gtdbtk.json.html">gtdbtk.json</a></li>
<li class="toctree-l2"><a class="reference internal" href="failed_genomes.tsv.html">failed.genomes.tsv</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html">Files</a> &raquo;</li>
        
      <li>summary.tsv</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/files/summary.tsv.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="summary-tsv">
<span id="files-summary-tsv"></span><h1>summary.tsv<a class="headerlink" href="#summary-tsv" title="Permalink to this headline">¶</a></h1>
<p>Classifications provided by the GTDB-Tk are in the files &lt;prefix&gt;.bac120.summary.tsv and &lt;prefix&gt;.ar53.summary.tsv for bacterial and archaeal genomes, respectively. These are tab separated files with the following columns:</p>
<ul class="simple">
<li><p>user_genome: Unique identifier of query genome taken from the FASTA file of the genome.</p></li>
<li><p>classification: GTDB taxonomy string inferred by the GTDB-Tk. An unassigned species (i.e., <code class="docutils literal notranslate"><span class="pre">s__</span></code>) indicates that the query genome is either i) placed outside a named genus or ii) the ANI to the closest intra-genus reference genome with an AF &gt;=0.65 is not within the species-specific ANI circumscription radius.</p></li>
<li><p>closest_genome_reference: indicates the accession number of the reference genome (species) to which a user genome was assigned based on ANI and AF. ANI values are only calculated when a query genome is placed within a defined genus and are evaluated for all reference genomes in that genus.</p></li>
<li><p>closest_genome_reference_radius: indicates the species-specific ANI circumscription radius of the reference genomes used to determine if a query genome should be classified to the same species as the reference.</p></li>
<li><p>closest_genome_taxonomy: indicates the GTDB taxonomy of the above reference genome.</p></li>
<li><p>closest_genome_ani: indicates the ANI between the query and above reference genome.</p></li>
<li><p>closest_genome_af: indicates the alignment fraction (AF) between the query and above reference genome.</p></li>
<li><p>closest_placement_reference: indicates the accession number of the reference genome when a genome is placed on a terminal branch.</p></li>
<li><p>closest_placement_taxonomy: indicates the GTDB taxonomy of the above reference genome.</p></li>
<li><p>closest_placement_ani: indicates the ANI between the query and above reference genome.</p></li>
<li><p>closest_placement_af: indicates the alignment fraction (AF) between the query and above reference genome.</p></li>
<li><p>pplacer_taxonomy: indicates the pplacer taxonomy of the query genome.</p></li>
<li><p>classification_method:        indicates the rule used to classify the genome. This field will be one of: i) ANI, indicating a species assignement was based solely on the calculated ANI and AF with a reference genome; ii) ANI/Placement, indicating a species assignment was made based on both ANI and the placement of the genome in the reference tree; iii) taxonomic classification fully defined by topology, indicating that the classification could be determine based solely on the genome’s position in the reference tree; or iv) taxonomic novelty determined using RED, indicating that the relative evolutionary divergence (RED) and placement of the genome in the reference tree were used to determine the classification.</p></li>
<li><p>note: provides additional information regarding the classification of the genome. Currently this field is only filled out when a species determination is made and indicates if the placement of the genome in the reference tree and closest reference according to ANI/AF are the same (congruent) or different (incongruent).</p></li>
<li><p>other_related_references: lists up to the 100 closest reference genomes based on ANI. ANI calculations are only performed between a query genome and reference genomes in the same genus.</p></li>
<li><p>msa_percent: indicates the percentage of the MSA spanned by the genome (i.e. percentage of columns with an amino acid).</p></li>
<li><p>red_value: indicates, when required, the relative evolutionary divergence (RED) for a query genome. RED is not calculated when a query genome can be classified based on ANI.</p></li>
<li><p>warnings: indicates unusual characteristics of the query genome that may impact the taxonomic assignment.</p></li>
</ul>
<div class="section" id="produced-by">
<h2>Produced by<a class="headerlink" href="#produced-by" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="../commands/classify.html#commands-classify"><span class="std std-ref">classify</span></a></p></li>
<li><p><a class="reference internal" href="../commands/classify_wf.html#commands-classify-wf"><span class="std std-ref">classify_wf</span></a></p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>user_genome classification  fastani_reference       fastani_reference_radius        fastani_taxonomy        fastani_ani     fastani_af      closest_placement_reference     closest_placement_taxonomy      closest_placement_ani   closest_placement_af    pplacer_taxonomy        classification_method   note    other_related_references(genome_id,species_name,radius,ANI,AF)  aa_percent      translation_table       red_value       warnings
genome_2    d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__VadinCA11 sp002498365    GCA_002498365.1 95.0    d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__VadinCA11 sp002498365    99.16   0.94    GCA_002498365.1 d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__VadinCA11 sp002498365    99.16   0.94    d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__ ANI/Placement   topological placement and ANI have congruent species assignments        GCA_002505345.1, s__VadinCA11 sp002505345, 95.0, 89.92, 0.89; GCA_002509405.1, s__VadinCA11 sp002509405, 95.0, 88.13, 0.89      87.1    11      N/A     N/A
genome_3    d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__VadinCA11 sp002498365    GCA_002498365.1 95.0    d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__VadinCA11 sp002498365    95.33   0.87    GCA_002498365.1 d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__VadinCA11 sp002498365    95.33   0.87    d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomethylophilaceae;g__VadinCA11;s__ ANI/Placement   topological placement and ANI have congruent species assignments        GCA_002505345.1, s__VadinCA11 sp002505345, 95.0, 94.26, 0.87; GCA_002509405.1, s__VadinCA11 sp002509405, 95.0, 90.74, 0.77      73.07   11      N/A     N/A
genome_1    d__Archaea;p__Euryarchaeota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanobrevibacter;s__Methanobrevibacter ruminantium     GCF_000024185.1 95.0    d__Archaea;p__Euryarchaeota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanobrevibacter;s__Methanobrevibacter ruminantium     100.0   1.0     GCF_000024185.1 d__Archaea;p__Euryarchaeota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanobrevibacter;s__Methanobrevibacter ruminantium     100.0   1.0     d__Archaea;p__Euryarchaeota;c__Methanobacteria;o__Methanobacteriales;f__Methanobacteriaceae;g__Methanobrevibacter;s__   ANI/Placement   topological placement and ANI have congruent species assignments        GCA_900321995.1, s__Methanobrevibacter sp900321995, 95.0, 80.9, 0.7; GCF_900114585.1, s__Methanobrevibacter olleyae, 95.0, 79.96, 0.55; GCA_900314635.1, s__Methanobrevibacter sp900314635, 95.0, 78.45, 0.3    97.09   11      N/A     N/A
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="test_execution.log.html" class="btn btn-neutral float-right" title="test_execution.log" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="red_dictionary.tsv.html" class="btn btn-neutral float-left" title="red_dictionary.tsv" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2025, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>