

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Change log &mdash; GTDB-Tk 1.4.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/custom.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="References" href="references.html" />
    <link rel="prev" title="Example" href="examples/classify_wf.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.4.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="announcements.html">Announcements</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing/index.html">Installing GTDB-Tk</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
</ul>
<p class="caption"><span class="caption-text">Running GTDB-Tk</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="commands/index.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption"><span class="caption-text">About</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Change log</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#id1">1.4.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id7">1.3.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id8">1.2.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id16">1.1.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id25">1.0.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id26">1.0.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id27">0.3.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id28">0.3.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id29">0.3.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id30">0.3.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id31">0.2.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id32">0.1.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id33">0.1.0</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Change log</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs_src/src/changelog.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="change-log">
<h1>Change log<a class="headerlink" href="#change-log" title="Permalink to this headline">¶</a></h1>
<div class="section" id="id1">
<h2>1.4.0<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Check if stdout is being piped to a file before adding colour.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/283">#283</a>) Significantly improved <code class="docutils literal notranslate"><span class="pre">classify</span></code> performance (noticeable when running trees &gt; 1,000 taxa).</p></li>
<li><p>Automatically cap pplacer CPUs to 64 unless specifying <code class="docutils literal notranslate"><span class="pre">--pplacer_cpus</span></code> to prevent pplacer from hanging.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/262">#262</a>) Added <code class="docutils literal notranslate"><span class="pre">--write_single_copy_genes</span></code> to the <code class="docutils literal notranslate"><span class="pre">identify</span></code> command. Writes unaligned single-copy AR122/BAC120 marker genes to disk.</p></li>
<li><p>Warn if GTDB-Tk is not running the most up-to-date version (disable via <code class="docutils literal notranslate"><span class="pre">GTDBTK_VER_CHECK</span> <span class="pre">=</span> <span class="pre">False</span></code> in <code class="docutils literal notranslate"><span class="pre">config.py</span></code>). If GTDB-Tk encounters an error it will silently continue (3 second timeout).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/276">#276</a>) Renamed the column <code class="docutils literal notranslate"><span class="pre">aa_percent</span></code> to <code class="docutils literal notranslate"><span class="pre">msa_percent</span></code> in <code class="docutils literal notranslate"><span class="pre">summary.tsv</span></code> (produced by <code class="docutils literal notranslate"><span class="pre">classify</span></code>).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/286">#286</a>) Fixed a file not found error when the reference data is a symbolic link (thanks <a class="reference external" href="https://github.com/davidealbanese">davidealbanese</a>!).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/277">#277</a>) Fixed an issue where if the user overrides the translation table using the optional 3rd column in the batchfile, the other coding density would appear as -100. Both translation table densities are now reported.</p></li>
<li><p>The <a class="reference internal" href="commands/check_install.html#commands-check-install"><span class="std std-ref">check_install</span></a> command now also checks that all third party binaries can be found on the system path.</p></li>
</ul>
</div>
<div class="section" id="id7">
<h2>1.3.0<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>This version of GTDB-Tk requires a new version of the GTDB-Tk reference package
(gtdbtk_r95_data.tar.gz) <a class="reference external" href="https://data.ace.uq.edu.au/public/gtdb/data/releases/release95/95.0/auxillary_files/gtdbtk_r95_data.tar.gz">available here</a>.</p></li>
<li><p>Updated reference package to use the <strong>GTDB Release 95</strong> taxonomy.</p></li>
<li><p>Report if the species-specific ANI circumscription criteria is satisfied in the <code class="docutils literal notranslate"><span class="pre">ani_closest.tsv</span></code> file output by <code class="docutils literal notranslate"><span class="pre">ani_rep</span></code>.</p></li>
<li><p>Estimated time until completion has been dampened.</p></li>
</ul>
</div>
<div class="section" id="id8">
<h2>1.2.0<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/241">#241</a>) Moved GTDB-Tk entry point to <code class="docutils literal notranslate"><span class="pre">__main__.py</span></code> instead of
<code class="docutils literal notranslate"><span class="pre">bin/gtdbtk</span></code> to support execution in some HPC systems (<code class="docutils literal notranslate"><span class="pre">gtdbtk</span></code> will still be aliased on install).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/251">#251</a>) Allow parsing of FastANI v1.0 output files. However, a warning will be displayed to update FastANI.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/254">#254</a>) Fixed an issue where <code class="docutils literal notranslate"><span class="pre">--scratch_dir</span></code> would fail, and not clean-up the mmap file.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/242">#242</a>) Added the <code class="docutils literal notranslate"><span class="pre">decorate</span></code> command allowing the <code class="docutils literal notranslate"><span class="pre">de</span> <span class="pre">novo</span> <span class="pre">workflow</span></code> to be run</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/244">#244</a>) Added the <code class="docutils literal notranslate"><span class="pre">infer_rank</span></code> method which established the taxonomic ranks of internal nodes of user trees based on RED</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/248">#248</a>) If the identify command is run on the same directory, genomes which were already processed will be skipped.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/pull/248">#248</a>) Improved <code class="docutils literal notranslate"><span class="pre">pplacer</span></code> output with running the <code class="docutils literal notranslate"><span class="pre">classify</span></code> command.</p></li>
</ul>
</div>
<div class="section" id="id16">
<h2>1.1.0<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>In rare cases pplacer would assign an empty taxonomy string which would raise an error.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/229">#229</a>) Genomes using windows line carriage <code class="docutils literal notranslate"><span class="pre">\r\n</span></code> would raise an error.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/227">#227</a>) CentOS machines would fail when using <code class="docutils literal notranslate"><span class="pre">~</span></code> in paths.</p></li>
<li><p>The bac120 symlink was pointing to the archaeal tree when using the <code class="docutils literal notranslate"><span class="pre">root</span></code> command.</p></li>
<li><p>Updated the <code class="docutils literal notranslate"><span class="pre">gtdb_to_ncbi_majority_vote.py</span></code> script for translating taxonomy.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/195">#195</a>) Added the <code class="docutils literal notranslate"><span class="pre">--pplacer_cpus</span></code> argument to specify the number of pplacer threads when running <code class="docutils literal notranslate"><span class="pre">classify</span></code> and <code class="docutils literal notranslate"><span class="pre">classify_wf</span></code> (#195).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/198">#198</a>) The <code class="docutils literal notranslate"><span class="pre">--debug</span></code> flag of <code class="docutils literal notranslate"><span class="pre">align</span></code> outputs aligned markers to disk before trimming.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/225">#225</a>) An optional third column in the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> will specify an override to which translation table should be used.
Leave blank to automatically determine the translation table (default).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/131">#131</a>) Users can now specify genomes which have NCBI accessions, as long as they are not GTDB-Tk
representatives (a warning will be raised).</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/191">#191</a>) Added a new command <code class="docutils literal notranslate"><span class="pre">ani_rep</span></code> which calculates the ANI of input genomes to all GTDB
representative genomes.</p></li>
<li><p>This command uses <a class="reference external" href="https://github.com/marbl/Mash">Mash</a> in a pre-filtering step. If pre-filtering is enabled (default)
then <code class="docutils literal notranslate"><span class="pre">mash</span></code> will need to be on the system path. To disable pre-filtering use the <code class="docutils literal notranslate"><span class="pre">--no_mash</span></code> flag.</p></li>
<li><p>(<a class="reference external" href="https://github.com/Ecogenomics/GTDBTk/issues/235">#230</a>) Improved how markers are used in determining the correct domain, and gene selection for the alignment.</p></li>
</ul>
</div>
<div class="section" id="id25">
<h2>1.0.2<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Fixed an issue where FastANI threads would timeout with <code class="docutils literal notranslate"><span class="pre">FastANI</span> <span class="pre">returned</span> <span class="pre">a</span> <span class="pre">non-zero</span> <span class="pre">exit</span> <span class="pre">code.</span></code></p></li>
<li><p>Versions affected: <code class="docutils literal notranslate"><span class="pre">1.0.0</span></code>, and <code class="docutils literal notranslate"><span class="pre">1.0.1</span></code>.</p></li>
</ul>
</div>
<div class="section" id="id26">
<h2>1.0.0<a class="headerlink" href="#id26" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Migrated to <strong>Python 3</strong>, you must be running at least <strong>Python 3.6</strong> or later to use this version.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">check_install</span></code> now does an exhaustive check of the reference data.</p></li>
<li><p>Resolved an issue where gene calling would fail for low quality genomes (#192).</p></li>
<li><p>Improved FastANI multiprocessing performance.</p></li>
<li><p>Third party software versions are reported where possible.</p></li>
</ul>
</div>
<div class="section" id="id27">
<h2>0.3.3<a class="headerlink" href="#id27" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>A bug has been fixed which affected <code class="docutils literal notranslate"><span class="pre">classify</span></code> and <code class="docutils literal notranslate"><span class="pre">classify_wf</span></code> when using the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>
argument with genome IDs that differed from the FASTA filename. This issue resulted in
the assigned taxonomy being derived only from tree placement without any ANI
calculations being considered. Consequently, in some cases genomes may have been classified as a new
species within a genus when they should have been assigned to an existing species. If you have genomes
with species assignments this bug did not impact you.</p></li>
<li><p>Progress is now displayed for: hmmalign, and pplacer.</p></li>
<li><p>Fixed an issue where the <code class="docutils literal notranslate"><span class="pre">root</span></code> command could not be run independently.</p></li>
<li><p>Improved MSA masking performance.</p></li>
</ul>
</div>
<div class="section" id="id28">
<h2>0.3.2<a class="headerlink" href="#id28" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>FastANI calculations are more robust.</p></li>
<li><p>Optimisation of RED calculations.</p></li>
<li><p>Improved output messages when errors are encountered.</p></li>
</ul>
</div>
<div class="section" id="id29">
<h2>0.3.1<a class="headerlink" href="#id29" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Pplacer taxonomy is now available in the summary file.</p></li>
<li><p>FastANI species assignment will be selected over phylogenetic placement (Topology case).</p></li>
</ul>
</div>
<div class="section" id="id30">
<h2>0.3.0<a class="headerlink" href="#id30" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Best translation table displayed in summary file.</p></li>
<li><p>GTDB-Tk now supports gzipped genomes as inputs (<code class="docutils literal notranslate"><span class="pre">--extension</span> <span class="pre">gz</span></code>).</p></li>
<li><p>By default, GTDB-Tk uses precalculated RED values.</p></li>
<li><p>New option to recalculate RED value during classify step (<code class="docutils literal notranslate"><span class="pre">--recalculate_red</span></code>).</p></li>
<li><p>New option to export the untrimmed reference MSA files.</p></li>
<li><p>New option to skip_trimming during align step.</p></li>
<li><p>New option to use a custom taxonomy file when rooting a tree.</p></li>
<li><p>New FAQ page available.</p></li>
<li><p>New output structure.</p></li>
</ul>
</div>
<div class="section" id="id31">
<h2>0.2.1<a class="headerlink" href="#id31" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Species classification is now based strictly on the ANI to reference genomes</p></li>
<li><p>The “classify” function now reports the closest reference genome in the summary file even if the ANI is &lt;95%</p></li>
<li><p>The summary.tsv file has 4 new columns: aa_percent, red_values, fastani_reference_radius, and warnings</p></li>
<li><p>By default, the “align” function now performs the same MSA trimming used by the GTDB</p></li>
<li><p>New pplacer support for writing to a scratch file (<code class="docutils literal notranslate"><span class="pre">--mmap-file</span></code> option)</p></li>
<li><p>Random seed option for MSA trimming has been added to allow for reproducible results</p></li>
<li><p>Configuration of the data directory is now set using the environment variable <code class="docutils literal notranslate"><span class="pre">GTDBTK_DATA_PATH</span></code> (see pip installation)</p></li>
<li><p>Perl dependencies has been removed</p></li>
<li><p>Python libraries biolib, mpld3 and jinja have been removed</p></li>
<li><p>This version requires a new version of the GTDB-Tk data package (gtdbtk.r86_v2_data.tar.gz) available <a class="reference external" href="https://data.ace.uq.edu.au/public/gtdbtk/release_86/">here</a></p></li>
</ul>
</div>
<div class="section" id="id32">
<h2>0.1.3<a class="headerlink" href="#id32" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>GTDB-Tk v0.1.3 has been released and addresses an issue with species assignments based on the placement of
genomes in the reference tree. This impacted species assignment when submitting multiple closely related genomes.
Species assignments reported by ANI were not impacted.</p></li>
</ul>
</div>
<div class="section" id="id33">
<h2>0.1.0<a class="headerlink" href="#id33" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Updated to R86, requires <a class="reference external" href="https://data.ace.uq.edu.au/public/gtdbtk/release_86/">release 86</a> to run.</p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="references.html" class="btn btn-neutral float-right" title="References" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="examples/classify_wf.html" class="btn btn-neutral float-left" title="Example" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Pierre-Alain Chaumeil, Aaron Mussig, and Donovan Parks

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>