

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Installing GTDB-Tk &mdash; GTDB-Tk 2.4.1 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  

  
  

  
    <link rel="canonical" href="https://ecogenomics.github.io/GTDBTk/installing/index.html" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <script defer data-domain="ecogenomics.github.io/gtdbtk" src="https://plausible.gtdb.ecogenomic.org/js/script.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>
    <script>
        window.dataLayer = window.dataLayer || [];

        function gtag() {
            dataLayer.push(arguments);
        }

        gtag('js', new Date());

        gtag('config', 'UA-84847737-2');
    </script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bioconda" href="bioconda.html" />
    <link rel="prev" title="Announcements" href="../announcements.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/GTDBTk.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.4.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../announcements.html">Announcements</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installing GTDB-Tk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sources">Sources</a><ul>
<li class="toctree-l3"><a class="reference internal" href="bioconda.html">Bioconda</a></li>
<li class="toctree-l3"><a class="reference internal" href="pip.html">pip</a></li>
<li class="toctree-l3"><a class="reference internal" href="docker.html">Docker</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#hardware-requirements">Hardware requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#python-libraries">Python libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="#third-party-software">Third-party software</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gtdb-tk-reference-data">GTDB-Tk reference data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Running GTDB-Tk</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../performance/index.html">Performance and Accuracy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands/index.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../files/index.html">Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/classify_wf.html">Example</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">GTDB-Tk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Installing GTDB-Tk</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/Ecogenomics/GTDBTk/blob/master/docs/src/installing/index.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="installing-gtdb-tk">
<span id="installing"></span><h1>Installing GTDB-Tk<a class="headerlink" href="#installing-gtdb-tk" title="Permalink to this headline">¶</a></h1>
<p>GTDB-Tk is available through multiple sources, you only need to choose one.
If you are unsure which one to choose, Bioconda is generally the easiest.</p>
<div class="section" id="sources">
<h2>Sources<a class="headerlink" href="#sources" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="bioconda.html">Bioconda</a></li>
<li class="toctree-l1"><a class="reference internal" href="pip.html">pip</a></li>
<li class="toctree-l1"><a class="reference internal" href="docker.html">Docker</a></li>
</ul>
</div>
<p>Alternatively, GTDB-Tk can be run online through <a class="reference external" href="https://kbase.us/applist/apps/kb_gtdbtk/run_kb_gtdbtk">KBase</a> (third party). Note that the version may not be the most recent release.</p>
</div>
<div class="section" id="hardware-requirements">
<h2>Hardware requirements<a class="headerlink" href="#hardware-requirements" title="Permalink to this headline">¶</a></h2>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Domain</p></th>
<th class="head"><p>Memory</p></th>
<th class="head"><p>Storage</p></th>
<th class="head"><p>Time</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Archaea</p></td>
<td><p>~60 GB</p></td>
<td><p>~140 GB</p></td>
<td><p>~90 minutes / 1,000 genomes &#64; 64 CPUs</p></td>
</tr>
<tr class="row-odd"><td><p>Bacteria</p></td>
<td><p>~110GB (690 GB when using –full_tree)</p></td>
<td><p>~140 GB</p></td>
<td><p>~90 minutes / 1,000 genomes &#64; 64 CPUs</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The amount reported of memory reported can vary depending on the number of pplacer threads.
See <a class="reference internal" href="../faq.html#faq-pplacer"><span class="std std-ref">GTDB-Tk reaches the memory limit / pplacer crashes</span></a> for more information.</p>
</div>
</div>
<div class="section" id="python-libraries">
<h2>Python libraries<a class="headerlink" href="#python-libraries" title="Permalink to this headline">¶</a></h2>
<p>GTDB-Tk is designed for Python &gt;=3.6 and requires the following libraries, which will be automatically installed:</p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 10%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Library</p></th>
<th class="head"><p>Version</p></th>
<th class="head"><p>Reference</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://dendropy.org/">DendroPy</a></p></td>
<td><p>&gt;= 4.1.0</p></td>
<td><p>Sukumaran, J. and Mark T. Holder. 2010. DendroPy: A Python library for phylogenetic computing. Bioinformatics 26: 1569-1571.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://numpy.org/">NumPy</a></p></td>
<td><p>&gt;= 1.9.0</p></td>
<td><p>Harris, C.R., Millman, K.J., van der Walt, S.J. et al. Array programming with NumPy. Nature 585, 357–362 (2020). DOI: <a class="reference external" href="https://doi.org/10.1038/s41586-020-2649-2">0.1038/s41586-020-2649-2</a></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://github.com/tqdm/tqdm">tqdm</a></p></td>
<td><p>&gt;= 4.35.0</p></td>
<td><p>DOI: <a class="reference external" href="https://doi.org/10.5281/zenodo.595120">10.5281/zenodo.595120</a></p></td>
</tr>
</tbody>
</table>
<p>Please cite these libraries if you use GTDB-Tk in your work.</p>
</div>
<div class="section" id="third-party-software">
<span id="installing-third-party-software"></span><h2>Third-party software<a class="headerlink" href="#third-party-software" title="Permalink to this headline">¶</a></h2>
<p>GTDB-Tk makes use of the following 3rd party dependencies and assumes they are on your system path:</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The <a class="reference internal" href="../commands/check_install.html#commands-check-install"><span class="std std-ref">check_install</span></a> command will verify that all of the programs are on the path.</p>
</div>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 10%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Software</p></th>
<th class="head"><p>Version</p></th>
<th class="head"><p>Reference</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="http://compbio.ornl.gov/prodigal/">Prodigal</a></p></td>
<td><p>&gt;= 2.6.2</p></td>
<td><p>Hyatt D, et al. 2010. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/20211023">Prodigal: prokaryotic gene recognition and translation initiation site identification</a>. <em>BMC Bioinformatics</em>, 11:119. doi: 10.1186/1471-2105-11-119.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="http://hmmer.org/">HMMER</a></p></td>
<td><p>&gt;= 3.1b2</p></td>
<td><p>Eddy SR. 2011. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/22039361">Accelerated profile HMM searches</a>. <em>PLOS Comp. Biol.</em>, 7:e1002195.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="http://matsen.fhcrc.org/pplacer/">pplacer</a></p></td>
<td><p>&gt;= 1.1</p></td>
<td><p>Matsen FA, et al. 2010. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/21034504">pplacer: linear time maximum-likelihood and Bayesian phylogenetic placement of sequences onto a fixed reference tree</a>. <em>BMC Bioinformatics</em>, 11:538.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://github.com/bluenote-1577/skani/">skani</a></p></td>
<td><p>&gt;= 0.2.1</p></td>
<td><p>Shaw J. and Yu Y.W. 2023. <a class="reference external" href="https://www.nature.com/articles/s41592-023-02018-3">Fast and robust metagenomic sequence comparison through sparse chaining with skani</a>. <em>Nature Methods</em>, 20, pages1661–1665 (2023).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a></p></td>
<td><p>&gt;= 2.1.9</p></td>
<td><p>Price MN, et al. 2010. <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2835736/">FastTree 2 - Approximately Maximum-Likelihood Trees for Large Alignments</a>. <em>PLoS One</em>, 5, e9490.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://github.com/marbl/Mash">Mash</a></p></td>
<td><p>&gt;= 2.2</p></td>
<td><p>Ondov BD, et al. 2016. <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x">Mash: fast genome and metagenome distance estimation using MinHash</a>. <em>Genome Biol</em> 17, 132. doi: doi: 10.1186/s13059-016-0997-x.</p></td>
</tr>
</tbody>
</table>
<p>Please cite these tools if you use GTDB-Tk in your work.</p>
</div>
<div class="section" id="gtdb-tk-reference-data">
<span id="installing-gtdbtk-reference-data"></span><h2>GTDB-Tk reference data<a class="headerlink" href="#gtdb-tk-reference-data" title="Permalink to this headline">¶</a></h2>
<p>GTDB-Tk requires ~140G of external data that needs to be downloaded and unarchived:</p>
<p><strong>For full package:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>https://data.ace.uq.edu.au/public/gtdb/data/releases/latest/auxillary_files/gtdbtk_package/full_package/gtdbtk_data.tar.gz
wget<span class="w"> </span>https://data.gtdb.ecogenomic.org/releases/latest/auxillary_files/gtdbtk_package/full_package/gtdbtk_data.tar.gz<span class="w"> </span><span class="o">(</span><span class="w"> </span>mirror<span class="w"> </span><span class="k">for</span><span class="w"> </span>Australia<span class="o">)</span>
tar<span class="w"> </span>xvzf<span class="w"> </span>gtdbtk_data.tar.gz
</pre></div>
</div>
<p><strong>For split package:</strong></p>
<p>To create an archive from the GTDB-Tk release data parts:</p>
<ol class="arabic simple">
<li><p>Ensure all parts of the GTDB-Tk release data are in the same directory.</p></li>
<li><p>Open a terminal or command prompt.</p></li>
<li><p>Navigate to the directory containing the parts of the GTDB-Tk release data.</p></li>
<li><p>Use the following command to concatenate all parts into a single archive:
cat gtdbtk_r220_data.tar.gz.part_* &gt; gtdbtk_r220_data.tar.gz</p></li>
<li><p>Once the command finishes executing, you will have a single archive file named ‘gtdbtk_r220_data.tar.gz’ in the same directory.</p></li>
</ol>
<p>You can find the gtdbtk_r220_data.tar.gz.part_* files under:
<a class="reference external" href="https://data.ace.uq.edu.au/public/gtdb/data/releases/release220/220.0/auxillary_files/gtdbtk_package/split_package/gtdbtk_r220_data.tar.gz.part_aa">https://data.ace.uq.edu.au/public/gtdb/data/releases/release220/220.0/auxillary_files/gtdbtk_package/split_package/gtdbtk_r220_data.tar.gz.part_aa</a></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that different versions of the GTDB release data may not run on all versions of GTDB-Tk, check the supported versions!</p>
</div>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 40%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>GTDB Release</p></th>
<th class="head"><p>Minimum version</p></th>
<th class="head"><p>Maximum version</p></th>
<th class="head"><p>MD5</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release220/226.0/auxillary_files/gtdbtk_package/full_package/gtdbtk_r226_data.tar.gz">R226</a></p></td>
<td><p>2.4.0</p></td>
<td><p>Current</p></td>
<td><p>24b476ea5a4ef30519d461e56cc4a27f</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release220/220.0/auxillary_files/gtdbtk_package/full_package/gtdbtk_r220_data.tar.gz">R220</a></p></td>
<td><p>2.4.0</p></td>
<td><p>Current</p></td>
<td><p>5aafa1b9c27ceda003d75adf238ed9e0</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release214/214.0/auxillary_files/gtdbtk_r214_data.tar.gz">R214</a></p></td>
<td><p>2.1.0</p></td>
<td><p>2.3.2</p></td>
<td><p>630745840850c532546996b22da14c27</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release207/207.0/auxillary_files/gtdbtk_r207_v2_data.tar.gz">R207_v2</a></p></td>
<td><p>2.1.0</p></td>
<td><p>2.3.2</p></td>
<td><p>df468d63265e8096d8ca01244cb95f30</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release207/207.0/auxillary_files/gtdbtk_r207_data.tar.gz">R207</a></p></td>
<td><p>2.0.0</p></td>
<td><p>2.0.0</p></td>
<td><p>b04c55104b491f84e053a9011b36164a</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release202/202.0/auxillary_files/gtdbtk_r202_data.tar.gz">R202</a></p></td>
<td><p>1.5.0</p></td>
<td><p>1.7.0</p></td>
<td><p>4986526c2b935fd4dcc2e604c0322517</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release95/95.0/auxillary_files/gtdbtk_r95_data.tar.gz">R95</a></p></td>
<td><p>1.3.0</p></td>
<td><p>1.4.2</p></td>
<td><p>06924c63f4b555ac6fd1525b09901186</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release89/89.0/gtdbtk_r89_data.tar.gz">R89</a></p></td>
<td><p>0.3.0</p></td>
<td><p>0.1.2</p></td>
<td><p>82966ef36086237d7230955e2bfff759</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release86/86.2/gtdbtk.r86_v2_data.tar.gz">R86.2</a></p></td>
<td><p>0.2.1</p></td>
<td><p>0.2.2</p></td>
<td><p>f71408d69fa2a289f2cdc734b7a58a02</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release86/86.0/gtdbtk_r86_data.tar.gz">R86</a></p></td>
<td><p>0.1.0</p></td>
<td><p>0.1.6</p></td>
<td><p>d019b3541746c3673181f24e666594ba</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://data.gtdb.ecogenomic.org/releases/release83/83.0/gtdbtk_r83_data.tar.gz">R83</a></p></td>
<td><p>0.0.6</p></td>
<td><p>0.0.7</p></td>
<td><p>9cf523761da843b5787f591f6c5a80de</p></td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="bioconda.html" class="btn btn-neutral float-right" title="Bioconda" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../announcements.html" class="btn btn-neutral float-left" title="Announcements" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2025, Pierre-Alain Chaumeil, Aaron Mussig and Donovan Parks.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>